path.MF.files="//Volumes/SONY HDD 1/Data/Project_224_Giorgetti_Boehm/"
path.MF.files.2="//Volumes/SONY HDD 1/Data/Project_758_Giorgetti_Boehm/"

# All files submitted to SRA, ID: PRJNA612865
# path.MF.files should be replaced with the folder containing files MF_TA_1i_R1.fastq.gz to MF_IgH_5ii_R2.fastq.gz *** being 100 in total *** 2x reads (R1, R2) 2x runs (i,ii) for 5x genes (TA,TB,TG,TD,IgH) for 5x fish (1 to 5)
# path.MF.files.2 should be replaced with a different folder containing files MF_1_R1.fastq.gz to MF_5_R2.fastq.gz *** being 10 in total *** 2 reads (R1, R2) of each of 5 fish (1 to 5), all genes under the same barcode for each fish.

# The following code outputs a list of data frames called MF.list to the working directory (plus a list of V genes to be used in the Markdown)
# The list has the following format MF.x.list[[gene]][[fish]]
# Names are added in the Markdown file (also present in the GitHub repository)

# The warnings at the end come from translating out of frame sequences and can be safely ignored
# Takes around 10 minutes to run in a regular computer

# IMPORTANT if needed, unload package seqinr before running as it contains a translate function that doesn't work with this code.
# To this end run the following line: detach("package:seqinr", unload=TRUE)

# IMPORTANT the Markdown file to be run after this code should be in the same folder as the output files
{
  inicio = Sys.time()
  # Libraries
  {
    library(ShortRead)
    library(tidyr)
    library(plyr)
    library(Biostrings)
    library(igraph)
  }
  # Databases
  
  {
    tr.DNA = function(st)translate(DNAStringSet(subseq(as.character(st),1,width=floor(nchar(as.character(st))/3)*3-3)),if.fuzzy.codon = "solve", no.init.codon = TRUE)
    tr.DNA3 = function(st)sapply(1:3,function(x)translate(DNAStringSet(subseq(st,x,width=floor(nchar(st)/3)*3-3)),if.fuzzy.codon = "solve", no.init.codon = TRUE))
    tr.DNA6 = function(st)return(unlist(list(forward=tr.DNA3(st),reverse=tr.DNA3(reverseComplement(DNAStringSet(st))))))
    
    # Database alpha
    {
      #CDR3.consensus.MF.a <- "SALYY|SAVYY|SAQYF|SAMYY|SGLYY|SALYF|AAVYY|SAKYY|SGLYF|INSKS|SVVYY|SALYH|SARYY"
      CDR3.consensus.MF.a <- "SALYY[CF]|SAVYYC|SAQYFC|SAMYYC|SGLYYC|SALYFC|AAVYYC|SAKYYC|SGLYFC|INSKSC|SVVYYC|SALYHC|SARYYC"
      J.consensus.MF.a <- "FG[A-Z]G[A,I,S,T]|FA[A-Z]G[I,T][T,K,F][L,V]|LG[S,T,F]GT|IASGTQL|FGK[Q,S]T|VGSGT"
      
      MF_Vas.5 =DNAStringSetList(Va632171 = "ATGGCCACAATTCCTTATATTCTGCTTCTGGTACTTCCGGTGGAAAGTCTCACATCCGCACAGACAATAATATCAATGGGAGAAGATACATTTGTAACAGAGGGAGATAAAGTCACTCTTCACTGTAGATACGATGGAGGAGTGAACAACCTACACTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATTGATTTATCCTGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAGGATAAAAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAGTGACTCTGCTCTGTACTACTGTGCACTGACGCCCACACTG",
                                 Va622124 = "ATGTCAGACTGGAAGCACAATGTCATCTTACTGCTCACAGGCCTTTGTGGATGTTATTGTGAGGAAAGCATTAACCAAAAAGAAAGAGACAAAACTGCCGTTGAAGGAGACAACGTTACAATCCACTGCACATATTCAACTAGTTCCACGTTTCCGTATTTTTTCTGGTATCAGCAGAAACTAAATGCGTCTCCTGTGTACATGCTGAACAAATATCTTGGAGCTATTGAAAAGGAAAGCGCATTTACAGAAAGATTTGATTCAAATCTCAAAAGCTCGTCCAACACAGTGTCACTGACAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGAGGCCCACAGTG",
                                 Va621304 = "ATGGCCACAATTCTTTATACTCTGCTTCTGATACTTCTGGTGATTAGTCTCACATCTGCACAGACAGTAACATCACTGGACGAAGAGACATTTGTAACAGAGGGGGATAAAGTCACTCTTCACTGTAGCTACGATGAAGGAGTGTACAACCTGCAGTGGTATCGCCAATATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTATTCCAGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAAGATAACAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAGTGACTCTGCTCTGTACTACTGTGCACTGACGCCCACACTG",
                                 Va604248.ps = "ATGTAGAAGAATCTTAACATTTCTCTTGATTTATAAAGCAGATACATTGACAGACTGGAAACGAAAATTAATTGTACTGTTCACAGCCCTGTGTGTGTTATTGTCAGGAAAACATCAACCAAAAAGAAAAACACAAAACTGCTGCTGAAGAGACGACGTTACAATCCACTTGCACATATTCAACTAGTAACAAGTTGCCGTATCTGTTCTGGTATCAGCGGACATTAAATACTCCTGTGTACATGCTGAAGTAATATCCTGGATCTATTGAAAAGGAAAGAGCATTTACAGAAAGATTTGATTCTGATCTCAACAGCTTGTCCAACACAGTGTCACTGACAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGAGGCCCACAGTG",
                                 Va603305 = "ATGGCCACAATTCCTTTTGTTCTGCTTCTCATACATCTGGTGGAAAGTCTCATATCCGCACAGACAATAACATCACTGGATGAAGAGACATTTGTAACAGAGGGAGATAAAGTCACTCTTTACTGTAGATACGATGGAGGAGTGAACAATCTGCAGTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTTTTCCAGCAGGAACAACCAGCAAACCTCTTCCACCTCGCTACGTACCAAAAGTTGATAAAGATAAAAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAATGACTCTGCTCTGTATTACTGTGCACTGACGCCCACACTG",
                                 Va589329 = "ATGACACGTCATTCACTTCTTCTGCTCTCTGCCCTGGTGTGTTCTACTTTAGGAGATGAAATCACTCCGGAGAAGACTGAAATAATATCTACAGAAGGATTAGCAGTAAAATTTACCTGCAACACTTCTGCATCAATTTACTCATTCCTCTGGTACCGTCAATATCCTGGATCAGCTCCTCAGTTTCTTGTGCTTCAGGTTGATGGAGTGAAAGGGATTAGGAAATCTGAAGTTGATCCCAGGTTCTCGACTATAATCAGAAAAGAAAAGCAGAGCTTTGTCGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGAGGCCCACAGTG",
                                 Va587123 = "ATGACAGTGTCAGGGACGGTTTCAGAAACGTCACAATAACTTTTACCATTCCTAAATCATCTATTTGAAATTGTATTTCACAAATTTCTCTTGCAGGTGTTTCTGCTGACTGCTGGATTAAACCAGATGGGAAGAAAAGGAGAGTCTGTGACGCTGAGATGCCCTTATTCCTCAAATAGTGATTATATTTATCTTTACTGGTACAGACAGTTTCCAAACACAGAACCTCAGTATTTACTCTATAAAGGTGCTAAAAGATGGAGTTCTACACACAGACTAAACAAACGATTTAGCACAATCACATCTGATTCTTCCACTGAAATTACCATTTACATTAACAGTAAGAGTTGCAGCACTACAAGCAACTAGAGAAGCTTAACAAAA",
                                 Va585213 = "ATGCTGGTCTTTCTATCTGTGCTGCTGGTTTTATTTAAAGGCAGACAAGCAGAAGATGTAATCTCACCATCCACAGAAGCAGCAGCTACATCAGAGAAGGAGAGTGTTACTCTCTCTTGTAATTACACTGGCTCAGTTGACAGTCTTCACTGGTATCGACAGAATGATGGTGAATCACCACAATTTCTAATACTTGACACTTATGGAACAGTTACTCAAGCCAGTCCTCCAGTTCCTGGATTGTCTATTCTTCATAGAAAAGAGGAGAAAAGTGTGGATCTGAAGATTTTTTCTGCTGCTGTATCAGACTCTGCTATGTATTACTGCGCTCTGCAGCCCACCGTG",
                                 Va584270 = "ATGGGTCTACACTCACTCCTTCTCTGTGCATCTATTGCCTTCACATTTGGAGATGTCATTACACCGAACAAAACGGATGTTTCAGCTGATGAAGGTTCCAGTGTGACTCTGTCTTGTATTTTCACTGCTACAGCATCAGATTATCTTCACTGGTATCGTCAACACGGAAGAACCAAACCTGAATTCCTTGTGCTGACATACAGCTTAGCAAAAGAAGCTCAACAATCTGACGTCGATCCAAGATTCTCAGTTAAAGTGAGAAAAGAGACCAGCTCTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGCAGCCCACAGTG",
                                 Va582747 = "ATGGCCACAATTCCTTATATTCTGCTTCTGATACTTCTGGTGATAAGTCTCACATTGTCTGCAGCACAGACAATAACATCACTGGACGAGGAGACATTTGTAACAGAGGGAGATAAAGTCACTCTTCACTGTAGCTACGATGGAGGAGTGAACAACCTTTTCTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTATTCCAGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAAGATAAAAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAGTGGCTCTGCTCTGTACTACTGTGCACTGACGCCCACACTG",
                                 Va580535 = "ATGGAGACAATTCTACTGCTGACTCTAATAACAGTGTCAGGTGTTTCTGCTGACTGGATTAAACCGGATGAGAAGAAGAGTGAAATCCGGAGAGAGGGAGAGTCAGTGACGCTGAGATGCTCTTATCAAACAAGCAGTGATTATATTTATCTTTACTGGTACAGACAGTTTACAAACACAGAACCTCAGTTTTTACTGAGCAAAGGGGAAAAAGGAAATAGTTATCAATATATACCAGACCAACGGTTTCACACGACCACATCTGATCATTCCACTGAAATCACCATTGACAGTGTAAATCTATCAGATTCAGCTCTGTATTTCTGCGCTCTTAGAGTTGCAGCACAGTG",
                                 Va571119 = "ATGTTGATATTTCTATTTGTGCTGCTGGTTTTATTTGAAGGCAGACAAGCAGAGGATGTAATCTCACCATCCACAGCAGCAGCAGCTACATCAGAGACGGAGAGTGTTACTCTCTCTTGTAATTACACTGGCTCAGTTGACAGTCTTCACTGGTATCGACAGAATGATGGTGAATCACCACAGTTTCTAATACTTGACTACTATGGAACAGTTACTCAAGCCAGTCCTCCAGTTCCTGGATTATCTATTCTTCATAGAAAAGAGGAGAAAAGTGTGGATCTGAAGATTGTTTCTGCAGCTGTATCAGACTCTGCTGTGTATTACTGCGCTCTGCAGCCCACCGTGACAGAAACCCCCCGCACCACACACAAAA",
                                 Va566811 = "ATGGAGACAATTCTACTGCTGACTCTAATAACAGTGTCAGGTGTTTCTGCTGACATGATTAAACCGAATGAGAAGAAGGCAATCCAGAGAGAGGGAGAGTCAGTGACGCTGAGATGCTCTTATTCTTCAAGCAGTGATTATATTGGTCTTCACTGGTACAGACAGTTTACAAACACAGAACCTCAATTTTTACTGCGGAAAGGGGCAAAAAGAAATAGTGGTCTATATATACCAGACCAACGGTTTCAGACGACCACATCTGATCATTCCACTGAAATCACCATTGACAGTGTGAATCTATCAGATTCAGCTCTGTATTTCTGCGCTCTAAGAGTTGCAGCACAGTG",
                                 Va564229 = "ATGGAAAATGGGCTTATGATCACTCTTCTCTTCATGACACTACATTGTGTGTGCTGGACACAAGACAAGGTTGAACAGAATTCAACAGAAATGACTGTGAAGGAACAACATCAAGTTGCTCTTTCTTGTAATTACTATACAACTGACACAAACCCATATCTCTTCTGGTACAAACAGCTGCCAAATGGGCCACCAAAATTTATTTTGAGTGAATTTACTTTTGGAAAAGGAGATACAGAGCCTGAATTTAAAGAGAGGTTCCATGCAACTCTGAACTCCACGTTGAAAAACATTCTTCTGGCAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGAGGCCCACAGTG",
                                 Va561593 = "ATGGATCTACAGTCTCTCATTCTTCTCTGTGCATCAGTTGCCTTCACATTTGGAGATGTCATTACACCGAGCAAAACGGAAGTTTCATCTTACGAAGGTTCCAATGTGACTTTGTCCTGTAATTTTACTGCTTCAGGTTCAGAATATCTTCTCTGGTATCGTCAACCTGGTAGAAACAAACCTGAATTTCTTGTGCTGACATTCAGAACAGCAAAAGACGCTCGAAAATCAGACGTCAATCCAAGATTCTCAGTTAAAGTGAGAAAAGAGAACAACTTTGTGGATCTGGAGATTTCTTCAGTTACGGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGAGGCCCACAGTGACAGGAAACACCACAACACTGTACAAAAACAC",
                                 Va559929 = "ATGTTCCTAATTCTGTGGTCAATATTACAACTTTGCTTTGGAACATCTTATGGTGATAAAATAGAAGCACTTTCATCAGTAGAACATGTTTTTGAAGGAGACACAGTTACTCTCTCGTGTAAATACAGCCTGACTTCTGGCACAAGTGCTACTCTACAGTGGTACAAACAGTCGCTCGGAACCAGACCAGAATACCTGTTGTTCGTCACAGAGTATTCAAACACATCTGATCCAGTTCTCCGAATGGTTTCAAGTGCTTCCAAAGAACTGAAGTGTGTGGATCTGGAGATTTCCTCTGTCACAGTGTCAGATTCTGCTCTATATCATTGCGCTCTGCAGCCCACAGTG",
                                 Va557552 = "ATGGCCACAATTCCTTATATTCTGTTTCTCATGTTTCTGATGGAAAGTCTCACATCTGCACAGACAATAACATCACTGGACGAAGAGATGCTTGTAACAGAGGGAGATAAAGTCACTCTTCACTGTAGCTACGATGGAGGAGTGAGCAACCTTTTCTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTATTCCTGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAAGATAAAAAACGTGTGAGTCTGGAGATCTTTACTTCAGAAGTCAGTGACTCTGCTCTGTACTACTGTGCACTGACGCCCACACTG",
                                 Va556665 = "ATGACAAAAAACTTGAAAATAAAATTAATTCTACTGTTCACAGCTCTGTCTGCTCCTTTAGGATGTTATTGTGAGGAAAGCATTAACCAAAAAGAAACAGACAAAACTGCTGTTGAAGGAGACGACGTTACGATCCACTGCACTTATTCAACCAGTGACACGTTTCCGTATCTGTTCTGGTATCAGCAGAAACTAAATGCGTCTCCTGTGTACATGCTTAAGAAATATCCTGGATCTATTGAAAAGGAAAGAGCATTTAGAGAAAGATTTGATTCAGATCTCAAAAGCTCGTCCAACACAGTGTCACTGACAATCCAGGAGGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGAGGCCCACAGTG",
                                 Va555641 = "ATGGAGACAATTCTACTGCTGACTCTAATAACAGTGTCAGGTGTTTCTGCTGACGTGATTAAACCGGATGAGAAGGCGGCAATCCAGAGAGAGGGAGAGTCAGTGACGCTGAGATGCTCTTATCAAACAAGCAGTGATTATATTTGTCTTCACTGGTACAGACAATTTACAAACACAGAACCTCAGTTTTTACTGAACAAAGGGGAAAAAGGAAATAGTTATCAATATATACCAAACCAACGGTTTCAGACGACCGCAGATGATTCTTTCACTGAGATCATCATCGATAGTGTAAATCTATCAGATTCAGCTCTGTATTTCTGCGCTCTAAGTTTTGCAGCACAGTG",
                                 Va551534 = "ATGGAAAATGGGCTTATGATCACTCTTCTCTTCATGACACTACATTATGTGTGCTGGACACAAGACAAGGTTGAACAGAATTTAACAGAAATGACTGTGAAGGAACAACATCAAGTTGCTCTTTCTTGTAATTACTATACAACTGACACAAACCCATATCTCTTCTGGTACAAACAGCTGCCAAATGGGCCACCAAAATTTATTTTGAGTGAATTTACTTTTGGAAAAGGAGATACAGAGCCTGAATTTAAAGAGAGGTTCCATGCAACTCTGAACTCCACGTTGAAAAACATTCTTCTGGCAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGCGCTCTGAGGCCCACAGTG",
                                 Va550438 = "ATGACAAAAAACTTGAAAACTAAATTAATTCTACTGTTCACAGCTCTGTCTGCTCCTTTAGGATGTTATTGTGAGGAAAGCATTAACCAAGAAGAAACGGACAAAACTGCTGTTGAAGGAGACGACGTTACGATCCACTGCACATATTCAACCAGTGACACGTTTCCGTATCTGTTCTGGTATCAGCAGAAACTAAATACGTCTCCTGTGTACATGCTTAAGAAATATCCTGGATCTATTGAAAAGGAAAGAGCATTTAGAGAAAGATTTGATTCAGATCTCAAAAGCTCGTCCAACACAGTGTCACTGACAATCCAGGAGGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGAGGCCCACAGTG",
                                 Va549599 = "ATGGATCTACAGTCTCTCATTCTTCTCTGTGCATCAGTTGCCGGCACATGTGGCAATGTCATTACACCGAAGGAAACTGAAGTTTCAGCTAATGAAGGTTCCAATGTGACTCTGTCTTGCAATTTTACTGCTTCAGGAGGTTCAGATGATCTGCTCTGGTATCATCGGCATGGAAAAACCAAACCTGAATTCATTGTACTGACATACAGAACAGCAAAAGATTCTAAACACTCTGACGTCGATTCAAGATTCTCAGTTAAAGTGAGAAAAGAGACCAGCTCTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGAGGCCCACAGTG",
                                 #Va548111 seems to be an artefactual duplication of 551534, not included, reverse orientation, two bp difference
                                 Va547224 = "ATGGCACGTCATTCACTTCTTCTGCTTTCTGCCCTGGTGTGTTCTGTTTTAGGAGATGAAATCACTCCTGAGAAGCCTGAGATGATTTCAGCCGAAGGTTCATCAGTAAAATTAACCTGCAACACTTCCTCTGCATCAGTAAACTCATTCCACTGGTACCGTCAATATCCAGGATCATCTCCTCAGTTTCTTGTTCTTCAAGTCGATGGTCTGAAAGGCATTAGGAAATCTGAAGTTGATCCCAGGTTCTCGACTCTAATCAGAAAAGAAAAGCAAAGCTTTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGCAGCCCACAGTG",
                                 Va545980 = "ATGACACGTCATTCACTCTTTTTGCTTTCTACCCTAGTGTGTTCTGTTTTAGGAGATGAAATCACTCCTGAGAAGACTGAGATGATTTATGCAGAAGGTTCAGAAGTAAAATTAACCTGCAACACTTCCTCTGCATCAATTTACTCATTCCTCTGGTACCGACAATATCCCGGATCAGCTCCTCAGTTTCTAGTTCTTCAGATGGCTGGCCTGAAAGACACTAAGAAATCTGAAGTTGATCCCAGGTTCTCGACTATAATCAGAAAGGAAACGCAAAGCTTTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGAGGCCCACAGTG",
                                 Va543308 = "ATGGCAAAATGGATGGCAATCGTTGTCTTCATATCACGAGCTTTAGTGTGCTGGGGGCAGAACATGGTTGAACAACCTTTAGGACAAAAGACATCGTTCGAAGCTGATACAGTTTCTTTAATGTGTAATTTTACCAGCACAACTACAAATTCTTATCTCTTCTGGTACAGAAAGTTGCCAAACAGAGCCCCAACTTTCATTTTGAGTGAATTTACTTTTGGAAAAGGAAATACTGGAGATAATTTCAAGAAAAGATTCTTCGCAAAATTGAATTCTACATTAAGAACCGTTCCCCTGACAATCAAGGATGTGCAAGTGTCTGACGCTGCTGTGTATTACTGCGCTCTGCAGCCCACCGTG",
                                 Va541109 = "ATGTTCATCTTTTTGTGGACAATACTGCAACTCTGCCTTGGATTCTGAACATCTGACGCTGATAAAATAGAAGCATTGTCGTCAGTAGAACATGTTTTTGAAGGAAGCACTGTTACTCTCTCGTGTAGTTACAGCCTGACTTCTGGCACAAGTGCTACCCTCCAATGGTACAAACAGTCGCTCGGAACCAGACCAGAATACCTGTTGTTCGTCACAGAGTATTCCAATACATCTGATCCAGTTCTCCGGATGGTTTCAAGTGCTTCCAAAGAACTGAAGCGTGTGGATCTGGAGATTTCCTCTGTTACAGTGTCAGATTCTGCTCTGTATTACTGCGCTCTGAGGCCCACAGCG",
                                 Va527192 = "ATGGATCTTCCTTCAGTTGTTCTTCTCTGTGCATCAGTTGCCGTCACGTTTGGAGATGTCATTACACCAAAGTGGACTGATGTCTCAGCAAATGAAGGTTCCAGTGTGACTCTGTCCTGCAATTTTACTGCTTCTGAAGTGTCTTCAGATTTTCTTTACTGGTATCGTCAACATGGAAGATCCAAACCTGAATTTCTTGTACAGACATACAGCTTAGCAAAAGAAGATCAACCATCTGAAGTCGATCCAAGATTCTCAGTTAAAGTGAGAAAAGAGAACAACTTTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGAGGCCCACAGTG",
                                 Va526060 = "ATGGATCTACAGTCTCTCATTCTTCTCTGCGCAATAGCTGCCGTCACGTTTGGAGATGTCATTACACCAAAGGAAACTGATGTTTCAGCTGATGAAGGTTCCAGTGTGACTCTGTCCTGTATTTATACTTCTGCAGATTACCTTCTCTGGTATCGTCAGCATGGAAGATCCAAGCCAGAATTTCTTGTAATAACAGTCAGAACAGCGAAAGAAGCTGAACAATCTGACGTCGATCCCAGATTCTCAGTTAAAGTGAGAAAAGAGACCAGCTCTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGAGGCCCACAGTG",
                                 Va523808 = "ATGGCGAATATCATGTTCTATCTTCTTATGGTGTCAGGTGTGTTATCATTGGACACAATTGCACCTCACAAAAGTCCTCTTATTTATATTGAACAAGGCAAAGTTAAAATGAGCTGTTCATATGAATCAACCAGTGATTATGTTTGGCTTTACTGGTTCAAGCAACAAAGCAACTCAGCACTTCAATATTTACGACATAAATCTGCACGATCGGCTAGTGGACAAGGAACTGATTCAGATCGTGTTCAATTGACTGCATCGCACGGATCCACAGAACTCAGCATTGACAGTGTAACTCTGTCAGATTCGGCTGTGTATTACTGTGCTTTACGATCCAGACGCCCAGTG",
                                 Va522454 = "ATGGCACGTCATTCGCTTTTTTTACTTTCTGCCATGGTGTGTTCTACTCTTGGAAATGATATCACTCCTGAAAAGACAGAAATTATTTCTACTGAAGGTTCATCAGTAAAATTAACCTGCAACACTTCCTCTGCATCAGTAAACTCATTCCTCTGGTACCGTCAATATCTAGGATCATCTCCTCAGTTTCTTGTGCTTCAGATTGATGGAGTGAAAGGCATTAGGGAATCTGAAGTTGATCCCAGGTTCTCGACTATAATCAGAAAAGAAAAGCAGAGCTCTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTGAGGCCCACAGTG",
                                 Va518791 = "ATGCCACCGAACAGTGCGAACACACAACCATGTATTTTTTTTCTTTTATTTCTGATTTCAGGTAGGACATTCTTCATGTGTTTCTTACATAAGTATATTTTGATTATCTTAATTCACCAATGTTTAAGTTTGTTTTTAGGTGTGATGACTGTGGACAAGATCTGGAGGAATGACGATTCTTTCAACAGAGCTGAACAGGAGGCTGTAACATTCAGATGTTCATATGAAACAAGCTCACAATATGTTTTACTTTACTGGTACATACAGTATCCCAACGGAGAACTTCGGTTTTTATTATTTAAACCAGGACGATCAGCCAGTTACAGTGAAAATCCTTCAGATCCTCGTCTAGAGTCGGACACAACTCACACCTCCAATACGCTGACCATCACAGCTTTATCTCTGTCAGACTCTGCTCTCTATTACTGCGCTCTAAGAACTGGTGCCCAGTG",
                                 Va510346 = "ATGGAAAAAGGCTACTCATTGATTTTCATAATAATTTCAGGTGTGTTTGCCAACAGCATTTGGCCGGCGGAGAAAGATACACACATATTTAAAACGGAGGGAGAGTCTGTGACAATGACATGTTCATTTGATATGACCAAAAATTACTTTTACCTTTACTGGTACCGGCAGTTTTCAAACACAGAACCCGAGTATTTACTGCATAAAGGTGCTCGTTTATTGAAAGACAGTCAGAGCATGTCTGATAAGAGATTTAAGACAAGTACTTCTGAAACATCCACCGAACTCACTATTAACAGACTGAGTGTGTCAGACTCAGCTGTCTATTACTGTGCTTTAAGACTGGATCCCCAGTG",
                                 Va509325 = "ATGAAACTAATAAGGATTATAATTATGACATCAGCATTTCTTGGTTGCCATTCAGCTGACAGTGTGGTACAGACACTTCACTTTAATACTACTGTACATGAGGGGGAATTCATCACACTGAACTGCACGTTTACCACCACTTCGACCACCCCAGAGCTTTTCTGGTATGTAAAAAGAAGAAACGATTCTCCTAAACTCATCTTGAGAAGAAATTCCTATGGAGGAGGACTGACCGAAGCAGAATTCCACAAGAGATTTGACTCTGAAGTCCAACAGTACAATTCAGTTCCTTTGATGATCCGAGACCTTCTTCTGTCAGACTCGGCTGTGTACTACTGTGCACTGAAGCCCACCATG",
                                 Va508727 = "ATGAACGTGGAAACATTTCTGCAACTTGTTCTTTTGATAGCAACAGGTGTGTTTTCTGACAGAATTGAACCAACAGACAAAAATACTAATGTGAAGAAAGGCGAAGGAGAGTCGGTGATGTTAAGCTGCACATACAGCACAATCAGCAATGATGTTTGGCTCTACTGGTACCGACAGTATTCTAACGCAGAACCTCAGTATTTACTGTGGAAAGGTGCACGATCATGGAACCAAGAGATGGAAATTCCAGACAAGAGATTCACGTCAACCACATCTCAATTTTCTGCTTCATTGACCATCGATTCTTTATCCCTCTCTGATTCAGCGCTTTACTACTGCGCTCTTCGAACCGCACGGCCCAGTG",
                                 Va507348 = "ATGTCGTATAGGAATATATGTTTTTTTATATTTTTGACATATTTCTGTGAATGTTACTGTCAAGACAGTGTGGACCAGCCAGAAAGAATTCAACACACAGTAGAAGATGTATCTGTAACGATAAACTGTGATTACAAAACCAGTGATTTATTGCCATATTTTTACTGGTACCAACAAAAAGCAAGTGGACCAATTCAACACATTATGACAATATTTGGAACATCTGTGCACAAGGAAAACACATTTAAGAAGAGATTCGATGCACATCACCATAAATTATCAAAATCAGTCCCGCTGACTATCCAGGATGTGCGAGTGTCAGACTCCGCTGTTTACTACTGTGCAATGCAGCCCACCATG",
                                 #Va506268.322 = "ATGTCCCTTATCTTAAGTGATTCAAAGCAGCCAGCAATTGTCCGAGTATGAAACATGTACATAACCTTGAGAAGCAGCGTTCGTTTTTTTTTTCTGACACTCTTTGCAACTGGCACTGAATGCAGAATACAAGACAACGTCCACCAGCCTGAACAGTTCCAGTCAGCGTTTGAAACGGGCCCAGTAACTTTGACATGCCAATATAACACTACAAGCTCAAACCCAACTCTGCTGTGGTACAGACAAGACAATTACTACTCTCCCAGGTATATTCTCCGGAGACATACGTTTGGAGACGGAGATGAAGCGGCGGAATTCAAGGGAAGATTTCAATCTAAACATAATCCAGTCACCAGTTCAGTTCCATTAACAATCCTGGAGCGAGTATCCGACTCTGCTGTTTACTACTGTGCTTTGCAGCCCACAGTG",
                                 Va506268 = "ATGTACATAACCTTGAGAAGCAGCGTTCGTTTTTTTTTTCTGACACTCTTTGCAACTGGCACTGAATGCAGAATACAAGACAACGTCCACCAGCCTGAACAGTTCCAGTCAGCGTTTGAAACGGGCCCAGTAACTTTGACATGCCAATATAACACTACAAGCTCAAACCCAACTCTGCTGTGGTACAGACAAGACAATTACTACTCTCCCAGGTATATTCTCCGGAGACATACGTTTGGAGACGGAGATGAAGCGGCGGAATTCAAGGGAAGATTTCAATCTAAACATAATCCAGTCACCAGTTCAGTTCCATTAACAATCCTGGAGCGAGTATCCGACTCTGCTGTTTACTACTGTGCTTTGCAGCCCACAGTG",
                                 Va505111 = "ATGATTCTCTGTTTTCTGATATTGCTCTCTACACTTTTACATGAGAGTAATGGAAGTTCCATCGCATCAGTCGACCCTGAGGTTTCTGGAGAAGAAGGGGAGAGTATAACTCTGTCCTGCAGCTACTCATCTGCTGTCACTTTGCTGTGGTACCGACAACACACAGACTCGGCTCCAGAACACCTTTTATCTATGTTGCATGTGACCGGCGAAGTTTTACAAAAGTCTGAGAAAATTGATCAAGAGTCTCGAATTTCTGGAAGACTAAACGAGAAGTCAACCCATACTGTTCTGGAGATTTCCTCCGCTCGAGTGTCAGATTCGGGTCTGTACTACTGTGCCATGGAGCCCACAGTG",
                                 Va502886 = "ATGATTTGTGTTTTGGGACTACTCTGTTCTCTGACTTATGTCAGTTTGGGGAATGTGATCACTCCGTTGGAGACACAGATGTCTGGGACTGAGGGAGATGTTTTAAGTTTCTCCTGTAACTATTCCTCAGCAGTTACTCTGTTCTGGTATCGCCAGTATCCTAAACTAGCTCCAGAATTTCTGATTAGTGTTCTACACTCTACAAGAAAAGTTCTTCAAAGATCTGAATTATTGGATGAAAACCAGATCTCTTCAAGTCTGAATGAACTGAAAACCCAAGTGTTTCTGGAGATTTCCTCTGCTCGAGTGTCAGATTCTGCTGTGTACTACTGTGCTGTGGAGCCCACAGTG",
                                 Va500782 = "ATGTTCAAAATGTATGTCTATGTAATAATATTGCTGTCCTTACATAAAGGTTCCAGAGCAGATATCATCACACCAAATCGTGACAGAGAATCTGGACTTGAGGGCGACAATGTGACTCTCTCTTGTAACTACAGTGATTCTGCTCGTGCTGTGCACTGGTATCGTCAGAGCGGAGCTTTTCAAATTATGTTTCTGATACTGGAAGATTCTGGGGTGTCTAGATCAGCAATTCCACCCAATCCTGGAATTTCTATTAGACACAACAAAGAAGAAAGTCGAGTGGATCTGGAGATCTCTAATGCTGAAGTGTCGGACTCTGCTGTTTACTACTGCGCTCTGAAGCCCACAGTG",
                                 #Va499295.364 = "ATGTCAAACTTCTCTCAGAGACACCAGATGCTCAGGAGAGCAGAAGCCTGGTCTTTTCTCAATTAAACAATGTTGACACATCTATTGGTTATGCTGGTTTTATCTGAAGCGAGACAAGCAGAAGATGTAATCTCACCATCCAAAGAAGCAGCAGCTACAACAGAGACAGAGAATGTTACTCTCTCTTGTAAATACACTGGTGCAGCTAAAATCCTTTACTGGTATCGGCAGAATGATGGTGGATCACTGAAATTTCTAATACTAGACTATTATGGAGCAGTTTCAAAAGCCAATCCTCCAGTTCCTGGATTGTTTATTCTTCATAGAAAAGAAGAGAAAAGTGTGGATCTGAAGATTGTTTCTGCTGCTGTATCAGACTCTGTTGTGTATTACTGCGCTCTGCAGCCCACCGTG",
                                 Va499295 = "ATGTTGACACATCTATTGGTTATGCTGGTTTTATCTGAAGCGAGACAAGCAGAAGATGTAATCTCACCATCCAAAGAAGCAGCAGCTACAACAGAGACAGAGAATGTTACTCTCTCTTGTAAATACACTGGTGCAGCTAAAATCCTTTACTGGTATCGGCAGAATGATGGTGGATCACTGAAATTTCTAATACTAGACTATTATGGAGCAGTTTCAAAAGCCAATCCTCCAGTTCCTGGATTGTTTATTCTTCATAGAAAAGAAGAGAAAAGTGTGGATCTGAAGATTGTTTCTGCTGCTGTATCAGACTCTGTTGTGTATTACTGCGCTCTGCAGCCCACCGTG",
                                 #Va499295.274 = "ATGCTGGTTTTATCTGAAGCGAGACAAGCAGAAGATGTAATCTCACCATCCAAAGAAGCAGCAGCTACAACAGAGACAGAGAATGTTACTCTCTCTTGTAAATACACTGGTGCAGCTAAAATCCTTTACTGGTATCGGCAGAATGATGGTGGATCACTGAAATTTCTAATACTAGACTATTATGGAGCAGTTTCAAAAGCCAATCCTCCAGTTCCTGGATTGTTTATTCTTCATAGAAAAGAAGAGAAAAGTGTGGATCTGAAGATTGTTTCTGCTGCTGTATCAGACTCTGTTGTGTATTACTGCGCTCTGCAGCCCACCGTG",
                                 Va497790 = "ATGAACGTATATTTTCTGCTGGTTTTTGTAATGTTACCAGAAGCATTTGCTGAAGGCATTGAACTGAAGAACGAGAACAGGAACATAGTTACAACAGAAGAAGAGTCTGTGATACTGAGCTGCTCATATTTCAGAAGCAGTGAAAAAAAGGTTTTTCATTTCTGGTACAGACAGTATCCAAACCACAATCCTCAATATTTAGTAGGAAAAGATGCTGGCTCAGCGAGTGAAAGTATTGAGGTTAAAGACTATCAGTTGGCTATATCTGATTCGTACAGTAACCTGACTATTAAAAACGCATCCTTGTCCGATTCAGCTCTTTATTACTTTGCTATCATAATCCAGTCTG",
                                 Va495588 = "ATGAAGTCATCTGTATTTCTTGCTTTGATTGCATTTATTGGTGCCAGTTTGCAGGACAGCATCACACCATTATCTCCATCTGTCTATGGAAACAAAGATGAAAAGGTGACAATTTCTTGCAAATATACCTACACTGTAGCCATGAACAACCTGCAGTGGTATCGCCAGTATCCCGAGTCCAGACCGGAATTCTTGGTTCTTGTCATGGAGTCAGGACAAAATCAGACTGCTGCAAATTTTCAGCCTCAACTTACAGCTAAAGTTGATAAAGGCAAGAAGATTTTGAGTCTGGATATTTTTCCTGCTGACTTGAAGGATTCAGCAGTATATTATTGTGCGCTGCAGCCCACAGTA",
                                 Va492241 = "ATGATCTCTCTGTTTTTTTTGCTTTTAGCTGTGCAAGATAAGTCCTCCGGCCAGTCAATAAACCCACATTGGAGCAATATTTCAACCACACAGGGAAGCACAGTCATTATCTCATGCAAATACGAAGGTCAAGCACGAAGTTTACATTGGTATCGGCAGAATCCGGGATCCAGACCGGAATTCCTTCTCCTGCAGTTTGAATCTGGAAGCACAACATATGCAGACCCTCCTTTCCCCAGAGTGAATGGCCGAGTGAATAAGATGGAGAAACAAGTTCATCTGGAGGTTTCTTCTGCCGCAGACACTGATTCAGCTCTGTACTTCTGTGCCCTGGAGCCCACACTG",
                                 Va491181 = "ATGATGTCTGTGTTCTGGCTGCTTGTGCTTGCAACTATGCAATACAGATGTTTCGCCCAGTCAATAACTGCGGTAAAGGACACAGTGCTCGTCACTGAGGGAAGTACAGTTCTACTATCATGCAAATATGATGGGAGTGCAAACAATTTGCAGTGGTATCGACAGTATCCTGGACAAAATCCTACATTTCTTCTTCTCATTTTTGAGTCTTCAACAAAGTCTCTTGTCCACGCATCTCCTCCAGTCCCAAGAGTGGATGCAGAAGTAAATGTGAAAGAAAAGTGTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGATTCTGCTCTATATTACTGCGCCCTGGTGCCCACAGTG",
                                 Va490153 = "ATGCTATTTGGTTGGCTGCCATTATTTTGGATTCTAGGACATTCCTTTGCTAATACAATACAACCATTCTTCACCGAGAAAAGAGTTTATGATGGCGAAAATTTGACTCTCTCTTGTAATTACTCTGGAATTGTTGACAACCTACAATGGTATCGTCAGTATTCTGGATCTTTCCCAGAGTTTCTGATATCCGTCACAGAATACACAGAAAACCGTGAGGCTTCAGGACGACTTTCATCAAAGGCAAAAAAAGACAAAAAGCGTGTGGATCTGGAGATTTCCTCCCTTACAGTGTCAGATTCTGCTCTGTATTACTGCGCCCTGGTGCCCACAGTG",
                                 Va489157 = "ATGATGCTGTTGTTTCCAACTATTTTAGCTTTTCTTGCAGTTTTAGAAAGCATTTCTGCAGATATGCTGATAAAGTCTGTCTCTTTGGAGCAGCGTGCACTAGAAGGGGTGACGGTTACTCTGACATGTAACTACACTGAAAGTAATGCGAGGAGTTATCAGTGGTACAGACAGCATCCCAACGCTGCTCCAGAGTTTCTCCTTCAAACCTTTGAAAGCACCGGATCACAGCAAAGCAATCGCTTCACTGCAGAGGCTCTAAAGAACAAGAAAGAGTTGCATCTGTCTGTATCCAGCACTGAAGGAGCTGATGCTGCCGTGTACTACTGTGCCTTGGTGCCCACAGTT",
                                 Va488009 = "ATGTTTGCGCTGATACTGAAGTTATGCTCAATATGTTTATCCTTCACAAAGACCTCAGAAGCATTTTCAATAGAACCTCTAACACCATCAAAGATGGTTTCCGCTGGACAATCTGCTGTCATCTCCTGTAATTACACTGGTCTAGGACCCAGTTTTCAGTGGTATCGCCAGTTTCCTGGATCCAGACCAGTGTTCCTGATTTTTCATACGGAGATTGGAGATTCTTCAGATCCCACTTTGCGCCTGCGGGCAGTGCCCAGGAAAAAGTCCTCACTAGTGGATCTGGAGATCAACTCGACTGAAGTGAAAGACTCGGCTGTGTATTACTGCGCTCTGACGACCACGCT",
                                 Va487342 = "ATGTATTTACATTGCTGGATGACACTGTTACTATTCATAGGACGGTCATCAACAGATGAAATTGAGCCCGTGTCTTCGGAAAAGAAAGTTTTTGAAGGACAAACAGTGACGCTGTCGTGTAACTACACTGGAACTGTGTACAATTTACTCTGGTATCGACAAAATGCGCAATCCAGACTGGACTTTCTGCTGATTGTGAGTGTCAACTCTGAAGCATCTGAACCTTCTCTACGTGTGTCTGTAAAGGACAGCAAGGTGGATAAACGAGTGGATCTGACGCTCTCTGCGGCTGAGGTGACAGACTCTGCTCTGTACTACTGTGCCCTGAGGCCCACACTG",
                                 Va485649 = "ATGTTGTACTGTACACTTATTTTATTCCCCTTTCTTGCGGGAGACTGCTTAGGGGATTTAATTACTTCTCATGGTTCTGAAAAAGAAACTCAAGATGGTAACAGTGCGAGTCTGTTCTGTAACTACAGTGGTGCTGTTTACAACCTACTGTGGTATCGTCAGTCCAGCGATGGAGGCAAACTACAATTTCTTCTCTCCGTCGCTGAATCTGGAGAGGCGAGTAAAGCTGACCCGTCAGCTCACTGGTTCAAACCCAAAGTGGACAAACACAACAGGTTAATGGAACTGGAAATCTCACCCAGTAAACTGACAGATTCTGCCAAATATTACTGTGCCCTGACAGCCACAGTA",
                                 Va483363 = "ATGCTTTTCTCGTTGTTCTTATTTTTATATTCCCTGCTCCGTGAGAATGCTGATGCAGATCTACTAAAACCACATTATCCAGAATATCAAGCTGTGGCAGGAGAAAACATTACTCTGTCTTGCAAATACAGTGGCCCTGTGAACTCACTGCAGTGGTATCGGCAAGTTCCAGGAGATAGGCTTAAATATCTGATTTTTACAACAGAACTCAGGAATCAGTCAGATTCTTCTCTTCGGCTTTTGGCTGTGGCTATTGAGAAAACCAAACAGATGTATTTGAGCATTTCTTCAGTAAAAGAGGAAGACTCCGGCTTGTATTTTTGCGCTCTTGTGGTGTCCACAGTGACAGGAAACACAGTCGTTCTTGACAAAAACTGC",
                                 Va479029 = "ATGTTACTCGAAATGTTTATACTTTTTATCAGTTTCTGTAAATCTTATGGAAACTTCATTGGACCAGACAAAACATCTCAAGTTATTACAGAGGGAAGCAGCACCAGACTGACATGTACATATAACGGATCTGCTTACAGTCTTCTGTGGTACCGACAAACACCAGGATCGAGACCAGAGTTTCTGCTGATGATTGCAGAGTCTTCAAAATACATCACAAAAGCTGAAACCCCTCACCCATCAATGTCCATTAAACTTCACCAGGATAAACGTGTGGATCTGGAGATTTCCTCTGCTTCAGTATCAGACTCGGCTCAGTATTTTTGTGCTCTGCAGCCCACAATGACAGGAAACTCAACAGAATGGCACAAAAACCTC",
                                 Va478395 = "ATGCGTACCATCATAGTTTATATCATTGCGTTTCTTCTAAAAGAGAGCGCTGAAGCAGATTTGATCAAATCATTCGCCTTGAAGCAGCAGATTGTGGCAGGAGAAGATGTTTCACTTTACTGTAACTACAGTGGGAGTGTACAAAATTTGCAGTGGTACCGTCAGTATCCTGGACTCAAACCTGAGAACTTAATTTTGTTCTTTGAGACTTTGCCCATGTCAGATCCGTCTCTTCGTCTTTCTGCTGCAGCTGAGAAATCAGTCAAGCAAATGAGTCTGACTATCTCCTCTACAGAAGTGTCAGACTCTGCTCTGTACTACTGTGCCCTGGAGCCCACAGTGACAGGAAACACAGCATCACTGTACAAAAACA")
      
      MF.Vas.observed = DNAStringSetList(Va632171.o = "ATGGCCACAATTCCTTATATTCTGCTTCTGGTACTTCCGGTGGAAAGTCTCACATCCGCACAGACAATAATATCAATGGGAGAAGATACATTTGTAACAGAGGGAGATAAAGTCACTCTTCACTGTAGATACGATGGAGGAGTGAACAACCTACACTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATTGATTTATCCTGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAGGATAAAAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAGTGACTCTGCTCTGTACTACTGTGCACTGACG",
                                         Va622124.o = "ATGTCAGACTGGAAGCACAATGTCATCTTACTGCTCACAGGCCTTTGTGGATGTTATTGTGAGGAAAGCATTAACCAAAAAGAAAGAGACAAAACTGCCGTTGAAGGAGACAACGTTACAATCCACTGCATATATTCAACTAGTTCCACGTTTCCGTATTTTTTCTGGTATCAGCAGAAACTAAATGCGTCTCCTGTGTACATGCTGAACAAATATCTTGGAGCTATTGAAAAGGAAAGCGCATTTACAGAAAGATTTGATTCAAATCTCAAAAGCTCGTCCAACACAGTGTCACTGACAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGAGGC",
                                         Va621304.o = "ATGGCCACAATTCTTTATACTCTGCTTCTGATACTTCTGGTGATTAGTCTCACATCTGCACAGACAGTAACATCACTGGACGAAGAGACATTTGTAACAGAGGGAGATAAAGTCACTCTTCACTGTAGCTACGATGAAGGAGTGTACAACCTGCAGTGGTATCGCCAATATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTATTCCAGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAAGATAACAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAGTGACTCTGCTCTGTACTACTGTGCACTGACG",
                                         Va603305.o = "ATGGCCACAATTCCTTTTGTTCTGCTTCTCATACATCTGGTGGAAAGTCTCATATCCGCACAGACAATAACATCACTGGATGAAGAGACATTTGTAACAGAGGGAGATAAAGTCGCTCTTTACTGTAGCTACGATGGAGGAGTGAACAATCTGCAGTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTTTTCCAGCAGGAACAACCAGCAAACCTCTTCCACCTCGCTACGTACCGAAAGTTGATAAAGATAAAAAACGTGTGAGTCTGGAGATCTTTCCATCTGAAGTCAGTGACTCTGCTCTGTACTACTGTGCACTG",
                                         Va589329.o = "ATGACACGTCATTCACTTCTTCTGCTCTCTGCCCTGGTGTGTTCTACTTTAGGAGATGAAATCACTCCGGAGAAGACTGAAATAATATCTACAGAAGGATTAGCAGTAAAATTTACCTGCAACACTTCTGCATCAATTTACTCATTCCTCTGGTACCGTCAATATCCTGGATCAGCTCCTCAGTTTCTTGTGCTTCAGGTTGATGGAGTGAAAGGGATTAGGAAATCTGAAGTTGATCCCAGGTTCTCGACTATAATCAGAAAAGAAAAGCAGAGCTTTGTCGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCT",
                                         Va585213.o = "ATGCTGGTCTTTCTATCTGTGCTGCTGGTTTTATTTAAAGGCAGACAAGCAGAAGATGTAATCTCACCATCCACAGAAGCAGCAGCTACACCAGAGACGAAGAGTATTACTCTCTCTTGTAATTACACTGGCTCAGTTAACAGTCTTCACTGGTATCGACAGAATGATGGTGAATCACCACAATTTCTAATACTTGACACTTATGGAACAGTTACTCAAGCCAGTCCTCCAGTTTCTGGATTATCTATTCTTCATCAAAAAGAAGAGAAAAGTGTGGATCTGAAGATTGTTTCTGCTGCTGTATCAGACTCTGCTGTGTATTACTGCGCTCTGCAGCC",
                                         Va584270.o = "ATGGGTCTACACTCACTCCTTCTCTGTGCATCTATTGCCTTCACATTTGGAGATGTCATTACACCGAACAAAACGGATGTTTCAGCTGATGAAGGTTCCAGTGTGACTCTGTCTTGTATTTTCACTGCTACAGCATCAGATTATCTTCACTGGTATCGTCAACACGGAAGAACCAAACCTGAATTCCTTGTGCTGACATACAGCTTAGCAAAAGAAGCTCAACAATCTGACGTCGATCCAAGATTCTCAGTTAAAGTGAGAAAAGAGACCAGCTCTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTG",
                                         Va582747.o = "ATGGCCACAATTCCTTATATTCTGCTTCTGATACTTCTGGTGATAAGTCTCACATTGTCTGCAGCACAGACAATAACATCACTGGACGAGGAGACATTTGTAACAGAGGGAGATAAAGTCACTCTTCACTGTAGCTACGATGGAGTCGTAAGCAACCTTTTCTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTATTCCAGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAAGATAACAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAGTGACTCTGCTCTGTATTACTGTG",
                                         Va571119.o = "ATGTTGATATTTCTATTTGTGCTGCTGGTTTTATTTGAAGGCAGACAAGCAGAGGATGTAATCTCACCATCCACAGCAGCAGCAGCTACATCAGAGACGGAGAGTGTTACTCTCTCTTGTAATTACACTGGCTCAGTTGACAGTCTTCACTGGTATCGACAGAATGATGGTGAATCACCACAGTTTCTAATACTTGACTACTATGGAACAGTTACTCAAGCCAGTCCTCCAGTTCCTGGATTATCTATTCTTCATAGAAAAGAGGAGAAAAGTGTGGATCTGAAGATTGTTTCTGCTGCTGTATCAGACTCTGCTGTGTATTACTGCGCCCTG",
                                         Va564229.o = "ATGGAAAATGGGCTTATGATCACTCTTCTCTTCATGACACTACATTATGTGTGCTGGACACAAGACAAGGTTGAACAGAATTCAACAGAAATGACTGTGAAGGAACAACATCAAGTTGCTCTTTCTTGTAATTACTATACAACTGACACAAACCCATATCTCTTCTGGTACAAACAGCTGCCAAATGGGCCACCAAAATTTATTTTGAGTGAATTTACTTTTGGAAAAGGAGATACAGAGCCTGAATTTAAAGAGAGGTTCCATGCAACTCTGAACTCCACGTTGAAAAACATTCTTCTGGCAATCCAGGAGGTGCGAGTGTGTGACTCTGCTGTGTATTACTGCGCTCTG",
                                         Va561593.o = "ATGGATCTACAGTCTCTCATTCTTCTCTGTGCATCAGTTGCCGTCACACTTGGAGATGTCATTACACCGAGCAAAACGGAAGTTTCATCTTACGAAGGTTCCAATGTGACTTTGTCCTGTAATTTTACTGCTTCAGGTTCAGAATATCTTCACTGGTATCGTCAACCTGGTAGAAACAAACCTGAATTTCTTGTGCTGACATTCAGAACAGCAAAAGACGCTCGAAAATCAGACGTCAATCCAAGATTCTCAGTTAAAGTGAGAAAAGAGAACAACTTTGTGGATCTGGAGATTTCTTCAGTTACGGTGTCAGACTCTGCTGTTTACTACTGC",
                                         Va559929.o = "ATGTTCGTAATTCTGTGGTCAATATTACAACTTTGCTTTGGAACATCTTATGGTGATAAAATAGAAGCACTTTCATCAGTAGAACATGTTTTTGAAGGAGACACAGTTACTCTCTCGTGTAAATACAGCCTGACTTCTGGCACAAGTCCTACTCTACAGTGGTACAAACAGTCGCTCGGAACCAGTCCAGAATACCTGTTGTTTGTCACAGAGTATTCCAACACATCTGATCCAGTTCTCCGGATGGTTTCAAGGGCTTCCAAAGAACTGAAGTGTGTGGATCTGGAGATTTCCTCTGTCACAGTGTCAGATTCTGCTCTATATCATTGCGCTCTGA",
                                         Va557552.o = "ATGGCCACAATTCCTTATATTCTGTTTCTCATGTTTCTGATGGAAAGTCTCACATCTGCACAGACAATAAAATCGCTGGACGAAGAGACATTTGTAACAGAGGGAGATAAAGTCACTCTTCACTGTAGCTACGATGGAGTCGTAAGCAACCTTTTCTGGTATCGCCAGTATCAAAGCTCTAATCCAGAGTTTCTGGCATGGATTATTCCTGGAGGAATGACCAGCAAACCTCTTCCACCTCGCTATGTACCGAAAGTTGATAAAGATAAAAAACGTGTGAGTCTGGAGATCTTTCCTTCTGAAGTCAGTGACTCTGCTCTGTACTACTGTGCACTGACG",
                                         Va556665.o = "ATGACAAAAAACTTGAAAATAAAATTAATTCTACTGTTCACAGCTCTGTCTGCTCCTTTAGGATGTTATTGTGAGGAAAGCATTAACCAAAAAGAAACAGACAAAACTGCTGTTGAAGGAGACGACGTTACGATCCACTGCACTTATTCAACCAGTGACACGTTTCCGTATCTGTTCTGGTATCAGCAGAAACTAAATGCGTCTCCTGTGTACATGCTTAAGAAATATCCTGGATCTATTGAAAAGGAAAGAGCATTTACAGAAAGATTTGATTCAGATCTCAAAAGCTCGTCCAACACAGTGTCACTGACAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGGA",
                                         Va551534.o = "ATGGAAAATGGGCTTATGATCACTCTTCTCTTCATGACACTACATTATGTGTGCTGGACACAAGACAAGGTTGAACAGAATTTAACAGAAATGACTGTGAAGGAACAACATCAAGTTGCTCTTTCTTGTAATTACTATACAACTGACACAAACCCATATCTCTTCTGGTACAAACAGCTGCCAAATGGGCCACCAAAATTTATTTTGAGTGAATTTACTTTTGGAAAAGGAGATACAGAGCCTGAATTTAAAGAGAGGTTCCATGCAACTCTGAACTCCACGTTGAAAAACATTCTTCTGGCAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGCGCTCTG",
                                         Va550438.o ="ATGACAAAAAACTTGAAAACTAAATTAATTCTACTGTTCACAGCTCTGTCTGCTCCTTTAGGATGTTATTGTGAGGAAAGCATTAACCAAGAAGAAACGGACAAAACTGCTGTTGAAGGAGACGACGTTACGATCCACTGCACATATTCAACCAGTGACACGTTTCCGTATCTGTTCTGGTATCAGCAGAAACTAAATACGTCTCCTGTGTACATGCTTAAGAAATATCCTGGATCTATTGAAAAGGAAAGAGCATTTAGAGAAAGATTTGATTCAGATCTCAAAAGCTCGTCCAACACAGTGTCACTGACAATCCAGGACGTGCGAGTGTGTGACTCTGCTGTGTATTACTGTGCTCTGAGG",
                                         Va549599.o = "ATGGATCTACAGTCTCTCATTCTTCTCTGTGCATCAGTTGCCGGCACATGTGGCAATGTCATTACACCGAAGGAAACTGAAGTTTCAGCTAATGAAGGTTCCAATGTGACTCTGTCTTGCAATTTTACTGCTTCAGGAGGTTCAGATGATCTGCTCTGGTATCATCGGCATGGAAAAACCAAACCTGAATTCATTGTACTGACATACAGAACAGCAAAAGATTCTAAACACTCTGACGTCGATTCAAGATTCTCAGTTAAAGTGAGAAAAGAGACCAGCTCTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTG",
                                         Va547224.o = "ATGGCACGTCATTCACTTCTTCTGCTTTCTGCCCTGGTGTGTTCTGTTTTAGGAGATGAAATCACTCCTGAGAAGCCTGAGATGATTTCAGCCGAAGGTTCATCAGTAAAATTAACCTGCAACACTTCCTCTGCATCAGTAAACTCATTCCACTGGTACCGTCAATATCCAGGATCATCTCCTCAGTTTCTTGTTCTTCAAGTCGATGGTCTGAAAGGCATTAGGAAATCTGAAGTTGATCCCAGGTTCTCGACTCTAATCAGAAAAGAAAAGCAAAGCTTTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTG",
                                         Va545980.o = "ATGACACGTCATTCACTCTTTTTGCTTTCTACCCTAGTGTGTTCTGTTTTAGGAGATGAAATCACTCCTGAGAAGACTGAGATGATTTATGCAGAAGGTTCAGAAGTAAAATTAACCTGCAACACTTCCTCTGCATCAATTTACTCATTCCTCTGGTACCGACAATATCCCGGATCAGCTCCTCAGTTTCTAGTTCTTCAGATGGCTGGCCTGAAAGACACTAAGAAATCTGAAGTTGATCCCAGGTTCTCGACTATAATCAGAAAGGAAACGCAAAGCTTTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTG",
                                         Va543308.o.1 = "ATGGCAAAATGGATGGCAATCGTTGTCTTCATATCACGAGCTTTAGTGTGCTGGGGGCAGAACATGGTTGAACAACCTTTAGGACAAAAGACATCGTTCGAAGCTGATACAGTTTCTTTAATGTGTAATTTTACCAGCACAACTACAAATTCTTATCTCTTCTGGTACAGAAAGTTGCCAAACAGAGCCCCAACTTTCATTTTGAGTGAATTTACTTTTGGAAAAGGAAATACTGGAGATAATTTCAAGAAAAGATTCTTCGCAAAATTGAATTCTACATTAAGAACCGTTCCCCTGACAATCCAGGATGTGCAAGTGTCTGACGCTGCTGTGTATTACTGC",
                                         Va543308.o.2 = "ATGGCAAAATGGATGGCAATCGTTGTCTTCATATCACGAGCTTTAGTGTGCTGGGGGCAGAACATGGTTGAACAACCTTTAGGACAAAAGACATCGTTCGAAGCTGATACAGTTTCTTTAATGTGTAATTTTACCAGCACAACTACAAATTCTTATCTCTTCTGGTACAGAAAGTTGCCAAACAGAGCCCCAACTTTCATTTTGAGTGAATTTACTTTTGGAAAAGGAAATACTGGAGATAATTTCAAGAAAAGATTCTTCGCAAAATTGAATTCTACATTAAGAACCGTTCCCCTGACAATCAAGGATGTGCAAGTGTCTGACGCTGCTGTGTATTACTGC",
                                         Va541109.o = "ATGTTCATCTTTTTGTGGACAATACTGCAACTCTGCCTTGGAACATCTGACGCTGATAAAATAGAAGCATTGTCGTCAGTAGAACATGTTTTTGAAGGAAGCACTGTTACTCTCTCGTGTAGTTACAGCCTGACTTCAAGTGCAAGTGCTACCCTCCAATGGTACAAACAGTCGCTCGGAACCAGACCAGAATACCTGTTGTTTGTCACAGAGTATTCCAATACATCTGATCCAGTTCTCCGGATGGTTTCAAGTGCTTCCAAAGAACTGAAGCGTGTGGATCTGGAGATTTCCTCTGTTACAGTGTCAGATTCTGCTCTGTATTACTGCGCTCTGACTACT",
                                         Va527192.o = "ATGGATCTTCCTTCAGTTGTTCTTCTCTGTGCATCAGTTGCCGTCACGTTTGGAGATGTCATTACACCAAAGTGGACTGATGTCTCAGCAAATGAAGGTTCCAGTGTGACTCTGTCCTGCAATTTTACTGCTTCTGAAGTGTCTTCAGATTTTCTTTACTGGTATCGTCAACATGGAAGATCCAAACCTGAATTTCTTGTACAGACATACAGCTTAGCAAAAGAAGATCAACCATCTGAAGTCGATCCAAGATTCTCAGTTAAAGTGAGAAAAGAGAACAACTTTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCT",
                                         Va526060 = "ATGGATCTACAGTCTCTCATTCTTCTCTGCGCAATAGCTGCCGTCACATTTGGAGATGTCATTGCACCAAAGGAAACTGATGTTTCAGCTGATGAAGGTTCCAGTGTGACTCTGTCCTGTAATTATACTTCTGCAGATTACCTTCTCTGGTATCGTCAGCATGGAAGATCCAAGCCAGAATTTCTTGTAATAACAGTCAGAACAGCGAAAGAAGCTGAACAATCTGACGTCGATCCCAGATTCTCAGTTAAAGTGAGAAAAGAGACCAGCTCTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTG",
                                         Va522454.o = "ATGGCACGTCATTCGCTTTTTTTACTTTCTGCCCTGGTGTGTTCTACTCTTGGAAATGATATCACTCCTGAAAAGACAGAAATTATTTCTACTGAAGGTTCATCAGTAAAATTAACCTGCAACACTTCCTCTGCATCAGTAAACTCATTCCTCTGGTACCGTCAATATCTAGGATCATCTCCTCAGTTTCTTGTGCTTCAGATTGATGGAGTGAAAGGCATTAGGGAATCTGAAGTTGATCCCAGGTTCTCGACTATAATCAGAAAAGAAAAGCAGAGCTTTGTGGATCTAGAGATTTCCTCAGTTACAGTGTCAGACTCTGCTGTTTACTACTGCGCTCTG",
                                         Va509325.o = "ATGAAACTAATAAGGATTATAATTATGACATCAGCATTTCTTGGTTGCCATTCAGCTGACAGTGTGGTACAGACACTTCACTTTAATACTACTGTACATGAGGGGGAATTCATCACACTGAACTGCACGTTTACCACCACTTCGACCACCCCAGAGCTTTTCTGGTATGTAAAAAGAAGAAACGATTCTCCTAAACTCATCTTGAGAAGAAATTCCTATGGAGGAGGACTGACCGAAGCAGAATTCCACAAGAGATTTGACTCTGAAGTCCAACAGTACAATTCAGTTCCTTTGATGATCCGAGACCTTCTTCTGTCAGACTCGGCTGTGTACTACTGTGCACTGAAGAATGCTGG",
                                         Va508727.o= "ATGAACGTGGAAACATTTCTGCAACTTGTTCTTTTGATAGCAACAGGTGTGTTTTCTGACAGAATTGAACCAACAGACAAAAATACTAATGTGAAGAAAGGCGAAGGAGAGTCGGTGATGTTAAGCTGCACATACAGCACAATCAGCAATGATGTTTGGCTCTACTGGTACCGACAGTATTCTAACGCAGAACCTCAGTATTTACTGTGGAAAGGTGCACGATCATGGAACCAAGAGATGGAAATTCCAGACAAGAGATTCACGTCAACCACATCTCAATTTTCTGCTTCATTGACCATCGATTCTTTATCCCTCTCTGATTCAGCGCTTTACTACTGCGCTCTTCGAACCGA",
                                         Va507348.o = "ATGTCGTATAGGAATATATGTTTTTTTATATTTTTGACATATTTCTGTGAATGTTACTGTCAAGACAGTGTGGACCAGCCAGAAAGAATTCAACACACAGTAGAAGATGTATCTGTAACGATAAACTGTGATTACAAAACCAGTTCTTTATTGCCATATTTCTACTGGTACCAACAAAAAGCAAGTGGACCAATTCAACACATTATGACAATATTTGGAACATCTGTGCATAAGGAAAACACATTTAAGAAGAGATTCGATGCACATCACCATAAATTATCAAAATCAGTCCAGCTGACTATCCAGGATGTGCGAGTGTCAGACTCCGCTGTTTACTACTGTGCAATG",
                                         Va506268.o = "ATGTACATAACCTTGAGAAGCAGCGTTCGTTTTTTTTTTCTGACACTCTTTGCAACTGGCACTGAATGCAGAATACAAGACAACGTCCACCAGCCTGAACAGTTCCAGTCAGCGTTTGAAACGGGCCCAGTAACTTTGACATGCCAATATAACACTACAAGCTCAAACCCAACTCTGCTGTGGTACAGACAAGACAATTACTACTCTCCCAGGTATATTCTCCGGAGACATACGTTTGGAGACGGAGATGAAGCGGCGGAATTCAAGGGAAGATTTCAATCTAAACATAATCCAGTCACCAGTTCAGTTCCATTAACAATCCTGGAGCGAGTATCCGACTCTGCTGTTTACTACTGTGCTTTGCAGAA",
                                         Va505111.o = "ATGATTCTCTGTTTTCTGATATTGCTCTCTACACTTTTACATGAGAGTAATGGAAGTTCCATCGCATCAGTCGACCCTGAGGTTTCTGGAGAAGAAGGGGAGAGTATAACTCTGTCCTGCAGCTACTCATCTGCTGTCACTTTGCTGTGGTACCGACAACACACAGACTCGGCTCCAGAACACCTTTTATCTATGTTGCATGTGACCGGCGAAGTTTTACAAAAGTCTGAGAAAATTGATCAAGAGTCTCGAATTTCTGGAAGACTAAACGAGAAGTCAACCCATACTGTTCTGGAGATTTCCTCCGCTCGAGTGTCAGATTCGGGTCTGTACTACTGTGCCATGGAG",
                                         Va502886.o = "ATGATTTGTGTTTTGGGACTACTCTGTTCTCTGACTTATGTCAGTTTGGGGAATGTGATCACTCCGTTGGAGACACAGATGTCTGGGACTGAGGGAGATGTTTTAAGTTTCTCCTGTAACTATTCCTCAGCAGTTACTCTGTTCTGGTATCGCCAGTATCCTAAACTAGCTCCAGAATTTCTGATTAGTGTTCTACACTCTACAAGAAAAGTTCTTCAAAGATCTGAATTATTGGATGAAAACCAGATCTCTTCAAGTCTGAATGAACTGAAAACCCAAGTGTTTCTGGAGATTTCCTCTGCTCGAGTGTCAGATTCTGCTGTGTACTACTGTGCTGTGGAG",
                                         Va500782.o.1 = "ATGTTCAAAATGTATGTCTATGTAATAATATTGCTGTCCTTACATAAAGGTTCCAGAGCAGATATCATCACACCAAATCGTGACAGAGAATCTGGACTTGAGGGCGACAATGTGACTCTCTCTTGTAACTACAGTGATTCTGCTCGTGCTGTGCACTGGTATCGTCAGAGCGGAGCTTTTCAAATTATGTTTCTGATACTGGAAGATTCTGGGGTGTCTAGATCAGCAATTCCACCCAATCCTGGAATTTCTATTAGACACAACAAAGAAGAAAGTCAAGTGGATCTGGAGATCTCTAATGCTGAAGTGTCGGACTCTGCTGTTTACTACTGCGCT",
                                         Va500782.o.2 = "ATGTTCAAAATGTATGTCTATGTAATAATATTGCTGTCCTTACATAAAGGTTCCAGAGCAGATATCATCACACCAAATCGTGACAGAGAATCTGGACTTGAGGGCGACAATGTGACTCTCTCTTGTAACTACAGTGATTCTGCTCGTGCTGTGCACTGGTATCGTCAGAGCGGAGCTTTTCAAATTATGTTTCTGATACTGGAAGATTCTGGGGTGTCTAGATCAGCAATTCCACCCAATCCTGGAATTTCTATTAGACACAACAAAGAAGAAAGTCGAGTGGATCTGGAGATCTCTAATGCTGAAGTGTCGGACTCTGCTGTTTACTACTGCGCT",
                                         Va499295.o = "ATGTTGACACATCTATTGGTTATGCTGGTTTTATCTGAAGCGAGACAAGCAGAAGATGTAATCTCACCATCCAAAGAAGCAGCAGCTACAACAGAGACAGAGAATGTTACTCTCTCTTGTAAATACACTGGTGCAGCTAAAATCCTTTACTGGTATCGGCAGAATGATGGTGGATCACTGAAATTTCTAATACTAGACTATTATGGAGCAGTTTCAAAAGCCAATCCTCCAGTTCCTGGATTGTCTATTCTTCATAGAAAAGAAGAGAAAAGTGTGGATCTGAAGATTGTTTCTGCTGCTGTATCAGACTCTGCTGTGTATGCTCTGCAGCC",
                                         Va495588.o = "ATGAAGTCATCTGTATTTCTTGCTTTGATTGCATTTATTGGTGCCAGTTTGCAGGACAGCATCACACCATTATCTCCATCTGTCTATGGAAACAAAGATGAAAAGGTGACAATTTCTTGCAAATATACCTACACTGTAGCCATGAACAACCTGCAGTGGTATCGCCAGTATCCCGAGTCCAGACCGGAATTCTTGGTTCTTGTCATGGAGTCAGGACAAAATCAGACTGCTGCAAATTTTCAGCCTCAACTTACAGCTAAAGTTGATAAAGGCAAGAAGATTTTGAGTCTGGATATTTTTCCTGCTGACTTGAAGGATTCAGCAGTATATTATTGTGCG",
                                         Va492241.o = "ATGATCTCTCTGTTTTTTTTGCTTTTAGCTGTGCAAGATAAGTCCTCCGGCCAGTCAATAAACCCACATTGGAGCAATATTTCAACCACACAGGGAAGCACAGTCATTATCTCATGCAAATACGAAGGTCAAGCACGAAGTTTACATTGGTATCGGCAGAATCCGGGATCCAGACCGGAATTCCTTCTCCTGCAGTTTGAATCTGGAAGCACAACATATGCAGACCCTCCTTTCCCCAGAGTGAATGGCCGAGTGAATAAGATGGAGAAACAAGTTCATCTGGAGGTTTCTTCTGCCGCAGACACTGATTCAGCTCTGTACTTCTGTGCCCTGGA",
                                         Va491181.o = "ATGATGTCTGTGTTCTGGCTGCTTGTGCTTGCAACTATGCAATACAGATGTTTCGCCCAGTCAATAACTGCGGTAAAGGACACAGTGCTCGTCACTGAGGGAAGTACAGTTCTACTATCATGCAAATATGATGGGAGTGCAAACAATTTGCAGTGGTATCGACAGTATCCTGGACAAAATCCTACATTTCTTCTTCTCATTTTTGAGTCTTCAACAAAGTCTCTTGTCCACGCATCTCCTCCAGTCCCAAGAGTGGATGCAGAAGTAAATGTGAAAGAAAAGTGTGTGGATCTGGAGATTTCCTCAGTTACAGTGTCAGATTCTGCTCTATATTACTGCGCCCTG",
                                         Va489157.o = "ATGATGCTGTTGTTTCCAACTATTTTAGCTTTTCTTGCAGTTTTAGAAAGCATTTCTGCAGATATGCTGATAAAGTCTGTCTCTTTGGAGCAGCGTGCACTAGAAGGGGTGACGGTTACTCTGACATGTAACTACACTGAAAGTAATGCGAGGAGTTATCAGTGGTACAGACAGCATCCCAACGCTGCTCCAGAGTTTCTCCTTCAAACCTTTGAAAGCACCGGATCACAGCAAAGCAATCGCTTCACTGCAGAGGCTCTAAAGAACAAGAAAGAGTTGCATCTGTCTGTATCCAGCACTGAAGGAGCTGATGCTGCCGTGTACTACTGTGCCTTG",
                                         Va488009.o = "ATGTTTGCGCTGATACTGAAGTTATGCTCAATATGTTTATCCTTCACAAAGACCTCAGAAGCATTTTCAATAGAACCTCTAACACCATCAAAGATGGTTTCCGCTGGACAATCTGCTGTCATCTCCTGTAATTACACTGGTCTAGGACCCAGTTTTCAGTGGTATCGCCAGTTTCCTGGATCCAGACCAGTGTTCCTGATTTTTCATACGGAGATTGGAGATTCTTCAGATCCCACTTTGCGCCTGCGGGCAGTGCCCAGGAAAAAGTCCTCACTAGTGGATCTGGAGATCAACTCGACTGAAGTGAAAGACTCGGCTGTGTATTACTGCGCTCTG",
                                         Va487342.o = "ATGTATTTACATTGCTGGATGACACTGTTACTATTCATAGGACGGTCATCAACAGATGAAATTGAGCCCGTGTCTTCGGAAAAGAAAGTTTTTGAAGGACAAACAGTGACGCTGTCGTGTAACTACACTGGAACTGTGTACAATTTACTCTGGTATCGACAAAATGCGCAATCCAGACTGGACTTTCTGCTGATTGTGAGTGTCAACTCTGAAGCATCTGAACCTTCTCTACGTGTGTCTGTAAAGGACAGCAAGGTGGATAAACGAGTGGATCTGACGCTCTCTGCGGCTGAGGTGACAGACTCTGCTCTGTACTACTGTGCCCTGA",
                                         Va485649.o = "ATGATGTTGTACTGTACACTTATTTTATTCCCCTTTCTTGCGGGAGACTGCTTAGGGGATTTAATTACTTCTCATGGTTCTGAAAAAGAAACTCAAGATGGTAACAGTGCGAGTCTGTTCTGTAACTACAGTGGTGCTGTTTACAACCTACTGTGGTATCGTCAGTCCAGCGATGGAGGCAAACTACAATTTCTTCTCTCCGTCGCTGAATCTGGAGAGGCGAGTAAAGCTGACCCGTCAGCTCACTGGTTCAAACCCAAAGTGGACAAACACAACAGGTTAATGGAACTGGAAATCTCACCCAGTAAACTGACAGATTCTGCCAAATATTACTGTGCCCTGA",
                                         Va483363.o = "ATGCTTTTCTCGTTGTTCTTATTTTTATATTCCCTGCTCCGTGAGAATGCTGATGCAGATCTACTAAAACCACATTATCCAGAATATCAAGCTGTGGCAGGAGAAAACATTACTCTGTCTTGCAAATACAGTGGCCCTGTGAACACACTGCAGTGGTATCGGCAAGTTCCAGGAGATAGGCTTAAATATCTGATTTTTACAACAGAACTCAGGAATCAGTCAGATTCTTCTCTTCGGCTTTTGGCTGTGGCTATTGAGAAAACCAAACAGATGTATTTGAGCATTTCTTCAGTAAAAGAGGAAGACTCCGGCTTGTATTTTTGCGCTCTTGTG",
                                         Va479029.o = "ATGTTACTCGAAATGTTTATACTTTTTATCAGTTTCTGTAAATCTTATGGAAACTTCATTGGACCAGACAAAACATCTCAAGTTATTACAGAGGGAAGCAGCACCAGACTGACATGTACATATAACGGATCTGCTTACAGTCTTCTGTGGTACCGACAAACACCAGGATCGAGACCAGAGTTTCTGCTGATGATTGCAGAGTCTTCAAAATACATCACAAAAGCTGAAACCCCTCACCCATCAATGTCCATTAAACTTCACCAGGATAAACGTGTGGATCTGGAGATTTCCTCTGCTTCAGTATCAGACTCGGCTCAGTATTTCTGTGCTCTGCAG",
                                         Va478395.o.1 = "ATGCGTACCATCATAGTTTATATCATTGCGTTTCTTCTAAAAGAGAGCGCTGAAGCAGATTTGATCAAATCATTCGCCTTGAAGCAGCAGATTGTGGCAGGAGAAGATGTTTCACTTTACTGTAACTACAGTGGGAGTGTACAAAATTTGCAGTGGTACCGTCAGTATCCTGGACTCAAACCTGAGAACTTAATTTTGTTCTTTGAGACTTTGCCCATGTCAGATCCGTCTCTTCGTCTTTCTGCTGCAGCTGAGAAATCAGTCAAGCAAATGAGTCTGACTATCTCCTCTACAGAAGTGTCAGACTCTGCTCTGTACTACTGCGCCCTG",
                                         Va478395.o.2 = "ATGCGTACCATCATAGTTTATATCATTGCGTTTCTTCTAAAAGAGAGCGCTGAAGCAGATTTGATCAAATCATTCGCCTTGAAGCAGCAGATTGTGGCAGGAGAAGATGTTTCACTTTACTGTAACTACAGTGGGAGTGTACAAAATTTGCAGTGGTACCGTCAGTATCCTGGACTCAAACCTGAGAACTTAATTTTGTTCTTTGAGACTTTGCCCATGTCAGATCCGTCTCTTCGTCTTTCTGCTGCAGCTGAGAAATCAGTCAAGCAAATGAGTCTGACTATCTCCTCTACAGAAGTGTCAGACTCTGCTCTGTACTACTGTGCCCTG")
      
      {
        MF_Vas.5.end=regexpr("SALYY|SAVYY|SAQYF|SAMYY|SGLYY|SALYF|AAVYY|SAKYY|SGLYF|INSKS|SVVYY|ITVL\\*G|SALYH",tr.DNA(unlist(MF_Vas.5)))
        MF_Vas.obs.end=regexpr("SALYY|SAVYY|SAQYF|SAMYY|SGLYY|SALYF|AAVYY|SAKYY|SGLYF|INSKS|SVVYY|ITVL\\*G|SALYH|SAVYAL",tr.DNA(unlist(MF.Vas.observed)))
        MF_Vas.5.end = MF_Vas.5.end*3
        MF_Vas.5.end[4] =MF_Vas.5.end[4]-10
        tr.DNA(subseq(unlist(MF_Vas.5),MF_Vas.5.end+13))
        MF_Vas.5.end.nt = subseq(unlist(MF_Vas.5),MF_Vas.5.end+13)
        MF.Vs.a.add = DNAStringSet(unlist(lapply(1:length(MF_Vas.5.end),function(i)paste0(c(as.character(MF_Vas.5.end.nt[i]),rep("A",36-nchar(MF_Vas.5.end[i]))),collapse = ""))))
        MF.Vas.last.36bp = subseq(MF.Vs.a.add,1,width = 36)
        names(MF.Vas.last.36bp) = subseq(names(MF_Vas.5),1,8)
        MF_Vas.5_toC = subseq(unlist(MF_Vas.5),1,MF_Vas.5.end+16)
      }
      
      MF.Js.a.2 = DNAStringSetList(MF.J.a.1 = "GTGAATCTCTGTGAGAATGCCATTGGATTCAGTCACGTCAGGTTTGGCCCAGGCATCA",
                                   MF.J.a.50 = "GGTAAATACTGTTGTGTGACTAGAGGAAATGACTATCTTAATTTTGCACAAGGAATCA", # new
                                   MF.J.a.51 = "GTGTAGTTGTGTGTGAATACTGGCATCCTGGGGAAAATAATATTTGCTCAGGGAATCA", # new
                                   MF.J.a.2 = "AACATTTCCCTTTGTGTGACTGGAAACATTAACCAGATCATTTTTGGGAAAGGAACGA", # old 2
                                   MF.J.a.3 = "GTGAGACACTGTGTGACTTCTGGAGGTTCGTGGAAGATCATTTTTGGAAAAGGAACAA",
                                   MF.J.a.61 = "GAGGTTTCACATTGTGTGACTAGAGGAAATTTTCAGATAATCTTTGGAACTGGCGTTG",  #new
                                   MF.J.a.4 = "AGAGAAACTCGGTGTGTGACAAGTGGAAATTGGAAAATCTATTTTGGCTCTGGGATCA", # old 4
                                   MF.J.a.5 = "GGTGTTACACGTTGTGTGACTGACAGTTATGGAAAAGTCATCTTTGGTTCTGGCACTA",
                                   MF.J.a.6 = "GATGACACTCACTGTGTGACTTATGGAATGGATAAAATCTATTTTGGCTCTGGAACAA",
                                   MF.J.a.7 = "AATGACACTTACTGTGTGACTTCTGGAATGAATAGAGTCTATTTTGGCTCTGGAACAA",
                                   MF.J.a.8 = "TAGATACTGTGTGTGACTCAGGCTGGCGGTTTGAAACTCATTTTTGGCTCTGGAACAA",
                                   MF.J.a.9 = "GTGTCCTAGCAATGTGTGAATACTGGAGCAAGAAAGATCGTTTTTGGAAAAGGAGCTC",
                                   MF.J.a.10 = "CTTAATCGTTGTGTGACTTCTACTGGATTTCAAAAAGTCATTTTTGGAACGGGCACAA",
                                   MF.J.a.52 = "GAAGCTGTATAATTGTGTGTGACTGGAATAGACAAACTGATATTTGCTGAAGGCACAA", # new
                                   MF.J.a.11 = "AACCCGTAGTGTGTGACTACTACTGGATTCAATAAGATTATATTTGGAAAAGGCACAA", # old 11
                                   MF.J.a.12 = "GATGCTTGACTGTGTGTGAATTTAGCTGGACAGAAAATTGTATTCGGTGGAGGCTCAA",
                                   MF.J.a.13 = "TTATACTGTGTAACTGATGGAAGTGGATTTAGAAAGTTGATATTTGGCTCAGGCACAA",
                                   MF.J.a.14 = "TGCTCTTTTGATGTGTGACGTAGTGGCTTGAGAAAGATTGTCTTTGGGACAGGCACAG",
                                   MF.J.a.53 = "GTAGATGTATAATTATGTGTGACTGTGGGACAGAAGATGATATTTGGAAAAAGCACAA", # new
                                   MF.J.a.15 = "CTGTTGTATTGTGTGAATGCAGCTGGTGGATATAAAATATTTTTTGGAACAGGAACGA", # old 15
                                   MF.J.a.16 = "CCGTTCCACTGTGTGACTACTGGAGTGGGATTTAAAATAATCTTCGGTACAGGCACCA",
                                   MF.J.a.17 = "TCATATTGTGTGACTACTGGAAGTGGACTTTATAAGCTTGTCTTTGGGAAAGGCACAG",
                                   MF.J.a.18 = "AAAATACAGTGTGTGACTCAGACTGGAGGGTGGAAAGTTCATTTTGGGTTTGGAACCA",
                                   MF.J.a.19 = "ATAAATTTTTTTTGAGAATGTGGACAAAAACAATTAGAAGTTTTTGGAAATGGCACCA",
                                   MF.J.a.54 = "GCTGTGAAGCTGTGTGTGACTAACTCTGGATATAAACTCGTACTGGGAACTGGCACCA", # new
                                   MF.J.a.20 = "GGTAGTCTTCACTGTGTGTCAACGAATTATGGAAAACTGATGTTTGGCCGGGGCACTA", # old 20
                                   MF.J.a.55 = "GTGTAGTACTGTGTGTCTGGAACTTTTGCTGGAAAAATCATTTTTGGAAAACAAACCA", # new
                                   MF.J.a.21 = "CAGCACGGTGTGAATTCAGGCAGTTCCTATGGAATACTCATCTTTGGGTCTGGAACAA", # old 21
                                   MF.J.a.22 = "AGGGTCTTACAATGTGTGACTACTGGAAATTACAGAATATATTTCGGCTCTGGATCAA",
                                   MF.J.a.23 = "AGTACACAACTTTGTGTGACTAACACAGTTGAATGGATAATATTTGGGGAAGGAACTA",
                                   MF.J.a.24 = "GTTTATAACTGTGTGAATACTGGAGGATATGGGAAGATAATCTTTGGAAAAGGAACTA", 
                                   MF.J.a.56 = "AGTGTGTGTTACTGTGTGACTTATTTAGCAGACAAAATCATATTTGCCACTGGAACAT", # new
                                   MF.J.a.25 = "GAAAATTCTAGTTGTGTGGATGGCAGCGGGAGAAAGATCATCTTTGGCCAAGGAACAG", # old 25
                                   MF.J.a.26 = "AGAGAAGTTTAGTGCTGTGTGAATGATGGCGCATGGAAAATCTTTGGATCTGGAACCA",
                                   MF.J.a.27 = "GAATGTTTTTACTGTGTGAATTACGCAACCAACCAAATCACTTTTGGACATGGAACAA",
                                   MF.J.a.28 = "GAATATTCTGGTTGTGTGGATGGCTTTGGGAGAAAGATCATCTTTGGCCAAGGAACCA",
                                   MF.J.a.29 = "GATGTTTAGTACTGTGTGAATGACGGTACAAGGAAAATCTTTTTTGGATCAGGAACTA",
                                   MF.J.a.30 = "TGTAGAGCTCGACAGTGTGTGAATTATGGCAACAAGATCATATTTGGGCATGGAACTC", # old 30
                                   MF.J.a.31 = "CATTTACCTTCTTACTGCGTGAATGCTGGGGTAAAGATAATATTTGGAACAGGAACTC",
                                   MF.J.a.32 = "GATCTTTTGCAATGTGTGAATGTCGGGGCTTACAAAATCATTTTTGGAAGAGGAACTC",
                                   MF.J.a.33 = "GCTGTGCTTATTCTGTGTGTGTATGCTGGGAACAAGATATTATTCGGTAGAGGAACTA",
                                   MF.J.a.34 = "GTCTCGTGTTATTGTGTGAATGATGGTAGAAGAAAAATTCTCTTTGGATCAGGGACCA",
                                   MF.J.a.35 = "TACTGTACAAGCAAACACTGTGTGAATAACGTGAAGATTGTATTTGGCAGTGGAACTA", # old 35
                                   MF.J.a.36 = "TCTGTAAGTCCTTACTGTGTGAGTTATGCAAACCAAATTATATTTGGCAAGGGAACAA",
                                   MF.J.a.37 = "TGTTCAACCAAACACTGTGTGAATGACGGATTGAGGATATTCTTTGGCAGTGGAACTA",
                                   MF.J.a.38 = "TGTACAACCAAACACTGTGTGAATACGGGGTTTAAGATTTTCTTTGGAAGTGGAACTA",
                                   MF.J.a.39 = "GACTGTGGCATATGTGTGACTCAGGGATATCAGAAAATTTCTTTTGGGTCTGGAACAA",
                                   MF.J.a.40 = "TGTTCAACTAAACACTGTGTGTATACTGGGTACAAGCTCTTCTTTGGCAGTGGAACTA", # old 40
                                   MF.J.a.41 = "AACAGCAGTTTGGAATGTGAATATGGACTAAACACGCTGATCTTTGGATTTGGGACAA",
                                   MF.J.a.42 = "TGTACATCCAAACACTGTGTGAATACAGGGATTAAAATCTTCTTTGGCAGTGGAACTA",
                                   MF.J.a.43 = "GACTGTAGTTGATGTGTGAATCAAGGATTTGGGAAAATTACTTTTGGATCTGGAACAA",
                                   MF.J.a.44 = "TGTTCAATCAAACACTGTGTGAATACTCAGTTCAAGATTTTCTTTGGCAGTGGAACTA",
                                   MF.J.a.45 = "GAATAAAGACAGTGTGTGAATCAAGGATTAAACAGGATGATCTTTGGATTTGGGACAA", # old 45
                                   MF.J.a.46 = "TGTACAACCAAACACTGTGTGAATACGGGACTTAATTTTGTCTTCGGCAGTGGAACTA",
                                   MF.J.a.57 = "GACTGTAGATGATGTGTGAATCTGGGAATTGGGAAAATTACTTTAGGATCTGGAACAA", # new
                                   MF.J.a.47 = "GTTTTCCACGTTTATTGTGTATATACTGGGAATATAATTTTCTTTGGCAGTGGATCTA", # old 47
                                   MF.J.a.48 = "TCCTCAACCAAACACTGTGTGAATAATGCCTATAAGATTTTCTTTGGCAGTGGAACTA",
                                   MF.J.a.49 = "TGTTCAACCAAACACTGTGTGAATGCTGGATATAAGATTTTCTTTGGCAGTGGAACTA",
                                   MF.J.a.58 = "TTGTATCCACTGAAATGCTGTGTGAATGTGAATAAGATCATCTTTGCTAAAGGAACGA", # new
                                   MF.J.a.59 = "GAACTTAGCCTTTGTGCGAATCTGGGAAATGACAAGCTGATGTTAGGATTTGGGACAA", # new
                                   MF.J.a.60 =  "GACTAAATTTAGTGTGTGAATCAAGGAAATTACAAACTCATTATTGCTTCTGGGACAC")
      
      
      MF.Js.a.aligned = as.character(unlist(MF.Js.a.2))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("TCACGTCAGGTTTGGC","TCACATCAGGTTTGGC",MF.Js.a.aligned[1]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("GGCGTTG","GGCATTA",MF.Js.a.aligned[6]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("CTGGAATGAATAGAG" ,"CTGGAATTAATAGAG",MF.Js.a.aligned[10]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("TTTTTGGAAAA","TTTTTGGGAAA",MF.Js.a.aligned[12]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("CTGGAGTGGGATTTAAAATAATCTTCGGT","CTGGAGCGGGATTTAAAATAATCTTCGGA",MF.Js.a.aligned[21]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub( "GAATGGA","GGACGGA",MF.Js.a.aligned[30]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("AGGAACTA","AGGAACTC",MF.Js.a.aligned[31]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("TGCTGTGTGAATGATGGCGCATGGAAAATCTTT", "TGTGTGAATGATGGCGCATGGAAAATCGTCTTT",MF.Js.a.aligned[34]))# variant to 26 (actually replaces 26) added triplet
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("ACAGGAACTC", "ACAGGAATTC",MF.Js.a.aligned[39]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("GATGGTAGAAG","GATGGCAGAAG",MF.Js.a.aligned[42]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("GAAGATTGTAT","GAAGATTTTAT",MF.Js.a.aligned[43]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("AGGATATTCTTT","AGGATATTCGTT",MF.Js.a.aligned[45]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("AAAATTTCTTTT","AAAATTACTTTT",MF.Js.a.aligned[47]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("AAGCTCTTCTTT","AAGCTATTCTTT",MF.Js.a.aligned[48]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("TAAAATCTTC","TAAAATTTTC",MF.Js.a.aligned[50]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("AAAATTACTTTT","AAAATTTATTTT",MF.Js.a.aligned[51])) #two nt changes added mar.9
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("TTGGGAAAATT","TTGGGAAAATG",MF.Js.a.aligned[55]))
      MF.Js.a.aligned = append(MF.Js.a.aligned,sub("TGGGACAAAACTAA","TGGTACAAAACTAC",MF.Js.a.aligned[60]))
      # MF.Js.a.aligned = append(MF.Js.a.aligned,sub("CTAAACGTG","CTAAACTTG",MF.Js.a.aligned[51])) dubious I can't find where this came from
      names(MF.Js.a.aligned) = c(1:length(MF.Js.a.aligned))
      MF.Ja.alleles.names = c(1,6,10,12,21,30,31,34,39,42,43,45,47,48,50,51,55,60)
      #MF.Ja.long is an extended version of the MF Ja until the splice donor
      MF.Ja.long = DNAStringSetList(MF.J.a.1 = "TGAGAATGCCATTGGATTCAGTCACGTCAGGTTTGGCCCAGGCATCAAACTCATAGTTGTGTCACGT",
                                    MF.J.a.50 = "GTGACTAGAGGAAATGACTATCTTAATTTTGCACAAGGAATCAAACTAAATATAGAGTCTGGT",
                                    MF.J.a.51 = "GTGAATACTGGCATCCTGGGGAAAATAATATTTGCTCAGGGAATCAAAGTTTATGTTCTTACACGT",
                                    MF.J.a.2 = "GTGTGACTGGAAACATTAACCAGATCATTTTTGGGAAAGGAACGAATGTGTTAATCGAAACAAGT",
                                    MF.J.a.3 = "TGTGACTTCTGGAGGTTCGTGGAAGATCATTTTTGGAAAAGGAACAAAATTGACAGTGAATTCAGGT",
                                    MF.J.a.61 = "TGACTAGAGGAAATTTTCAGATAATCTTTGGAACTGGCGTTGAAGTCAATATTCATACAGGT",
                                    MF.J.a.4 = "GTGACAAGTGGAAATTGGAAAATCTATTTTGGCTCTGGGATCAAACTGACAGTGGAAGCCAGT",
                                    MF.J.a.5 = "GTGACTGACAGTTATGGAAAAGTCATCTTTGGTTCTGGCACTAGATTACACATCAATGCAAGT",
                                    MF.J.a.6 = "GTGACTTATGGAATGGATAAAATCTATTTTGGCTCTGGAACAAAACTAAAGATTGAATCCCGT",
                                    MF.J.a.7 = "TGACTTCTGGAATGAATAGAGTCTATTTTGGCTCTGGAACAAAACTAAAGATTGAAACAGGT",
                                    MF.J.a.8 = "TGACTCAGGCTGGCGGTTTGAAACTCATTTTTGGCTCTGGAACAAAACTTAACGTTGAAAGCAGT",# not found
                                    MF.J.a.9 = "GTGAATACTGGAGCAAGAAAGATCGTTTTTGGAAAAGGAGCTCGATTAACAGTTCATTCAAGT",
                                    MF.J.a.10 = "GTGACTTCTACTGGATTTCAAAAAGTCATTTTTGGAACGGGCACAAAACTTGTTGTGAATGCAGGT",
                                    MF.J.a.52 = "GTGACTGGAATAGACAAACTGATATTTGCTGAAGGCACAAAACTCACTGTAAATGCAGGT",
                                    MF.J.a.11 = "GTGACTACTACTGGATTCAATAAGATTATATTTGGAAAAGGCACAAAACTGCTCATTAATCCAGGT",
                                    MF.J.a.12 = "GTGAATTTAGCTGGACAGAAAATTGTATTCGGTGGAGGCTCAAAACTAACTGTACATACAGGT",
                                    MF.J.a.13 = "TAACTGATGGAAGTGGATTTAGAAAGTTGATATTTGGCTCAGGCACAAATTTACATGTAAATGAAGGT",
                                    MF.J.a.14 = "TGACGTAGTGGCTTGAGAAAGATTGTCTTTGGGACAGGCACAGCGCTGACAGTCCAGACCAGT",
                                    MF.J.a.53 = "TGACTGTGGGACAGAAGATGATATTTGGAAAAAGCACAAGACTAATTATAAATGCAGGT",
                                    MF.J.a.15 = "GTGAATGCAGCTGGTGGATATAAAATATTTTTTGGAACAGGAACGAAGCTGAAAGTTGAAACAAGT",
                                    MF.J.a.16 = "GTGACTACTGGAGTGGGATTTAAAATAATCTTCGGTACAGGCACCAAACTTATGATCCAACCAAGT",
                                    MF.J.a.17 = "GTGACTACTGGAAGTGGACTTTATAAGCTTGTCTTTGGGAAAGGCACAGAATTGTTTGTTAAAACAAGT",
                                    MF.J.a.18 = "TGACTCAGACTGGAGGGTGGAAAGTTCATTTTGGGTTTGGAACCAAACTAACAGTTGAAAGTAGT",
                                    MF.J.a.19 = "GAATGTGGACAAAAACAATTAGAAGTTTTTGGAAATGGCACCATTTTTCCATTTCAAATTAGT", #not found, p
                                    MF.J.a.54 = "GTGACTAACTCTGGATATAAACTCGTACTGGGAACTGGCACCAAGCTCACAGTTCAAACAGGT",
                                    MF.J.a.20 = "TGTCAACGAATTATGGAAAACTGATGTTTGGCCGGGGCACTAGTGTACACGTTCAGATCGGT",
                                    MF.J.a.55 = "GTCTGGAACTTTTGCTGGAAAAATCATTTTTGGAAAACAAACCAAACTGCTGGTTTACTCCGGT",
                                    MF.J.a.21 = "TGAATTCAGGCAGTTCCTATGGAATACTCATCTTTGGGTCTGGAACAAGGTTAAGTGTTAGTACAAGT",
                                    MF.J.a.22 = "TGTGACTACTGGAAATTACAGAATATATTTCGGCTCTGGATCAAAACTTAACGTAGAAACAAGT",
                                    MF.J.a.23 = "GACTAACACAGTTGAATGGATAATATTTGGGGAAGGAACTAGACTTATTGTCAGTTCAGGT",  #not found
                                    MF.J.a.24 = "TGTGAATACTGGAGGATATGGGAAGATAATCTTTGGAAAAGGAACTAAACTTTTTGTACTGAAAAGT",
                                    MF.J.a.56 = "GTGACTTATTTAGCAGACAAAATCATATTTGCCACTGGAACATTTTTGGAAATTCATTCAAGT",
                                    MF.J.a.25 = "TGTGGATGGCAGCGGGAGAAAGATCATCTTTGGCCAAGGAACAGCTTTAAAAATTCAAACAAGT",
                                    MF.J.a.26 = "TGAATGATGGCGCATGGAAAATCTTTGGATCTGGAACCAAACTTTTCATAGAGCAAAGT",  #not found
                                    MF.J.a.27 = "TGTGAATTACGCAACCAACCAAATCACTTTTGGACATGGAACAATTTTAACAGTTGAATCAAGT",
                                    MF.J.a.28 = "GGATGGCTTTGGGAGAAAGATCATCTTTGGCCAAGGAACCATTTTAAAAGTTCAAACAAGT",
                                    MF.J.a.29 = "TGAATGACGGTACAAGGAAAATCTTTTTTGGATCAGGAACTAAACTTTTCATTGAGCAAAGT",
                                    MF.J.a.30 = "GTGAATTATGGCAACAAGATCATATTTGGGCATGGAACTCAACTAAATGTTTATTCTAGT",
                                    MF.J.a.31 = "TGAATGCTGGGGTAAAGATAATATTTGGAACAGGAACTCATCTTACAGTTCAAACTAGT",
                                    MF.J.a.32 = "TGTGAATGTCGGGGCTTACAAAATCATTTTTGGAAGAGGAACTCGAGTTTCAGTCCAGCCAAGT",
                                    MF.J.a.33 = "GTGTATGCTGGGAACAAGATATTATTCGGTAGAGGAACTAATCTCATTGTAACTACAGGT",
                                    MF.J.a.34 = "GAATGATGGTAGAAGAAAAATTCTCTTTGGATCAGGGACCAGACTTTTCATAGAACCAAGT",
                                    MF.J.a.35 = "TGAATAACGTGAAGATTGTATTTGGCAGTGGAACTAAACTTGTTGTAGAATCAAGT",
                                    MF.J.a.36 = "TGTGAGTTATGCAAACCAAATTATATTTGGCAAGGGAACAAGGCTAAATATTTATTCAAGT",
                                    MF.J.a.37 = "TGAATGACGGATTGAGGATATTCTTTGGCAGTGGAACTAAACTTGTTGTAGAATCAAGT",
                                    MF.J.a.38 = "GTGAATACGGGGTTTAAGATTTTCTTTGGAAGTGGAACTAAACTTGTTGTATTTCAAAGT",
                                    MF.J.a.39 = "GTGACTCAGGGATATCAGAAAATTTCTTTTGGGTCTGGAACAAAACTAACAGTTCTTTCAAGT",
                                    MF.J.a.40 = "GTGTATACTGGGTACAAGCTCTTCTTTGGCAGTGGAACTAAACTTGTTGTAGAACCAAGT",
                                    MF.J.a.41 = "TGAATATGGACTAAACACGCTGATCTTTGGATTTGGGACAAAAGTTTCTGTTTATTCAAGT", #not found
                                    MF.J.a.42 = "TGAATACAGGGATTAAAATCTTCTTTGGCAGTGGAACTAAACTTGTGGTAGATCAAAGT",
                                    MF.J.a.43 = "GTGAATCAAGGATTTGGGAAAATTACTTTTGGATCTGGAACAAAACTAACGGTTCTTTCAAGT",
                                    MF.J.a.44 = "TGAATACTCAGTTCAAGATTTTCTTTGGCAGTGGAACTAAACGTGTTGTAGAACCAAGT",
                                    MF.J.a.45 = "TGAATCAAGGATTAAACAGGATGATCTTTGGATTTGGGACAAAAGTTTCAGTTTTTTCAAGT",
                                    MF.J.a.46 = "TGAATACGGGACTTAATTTTGTCTTCGGCAGTGGAACTAAACTTGTTGTAGAAATAAGT",
                                    MF.J.a.57 = "TGAATCTGGGAATTGGGAAAATTACTTTAGGATCTGGAACAAGACTAACAGTTCTTTCAAGT",
                                    MF.J.a.47 = "GTATATACTGGGAATATAATTTTCTTTGGCAGTGGATCTAAACTTGTTGTAGAACCAAGT", #not found
                                    MF.J.a.48 = "TGAATAATGCCTATAAGATTTTCTTTGGCAGTGGAACTAAACTTCTAGTACAACCAAGT",
                                    MF.J.a.49 = "GTGAATGCTGGATATAAGATTTTCTTTGGCAGTGGAACTAAACTTGTTGTTGAATCAAGT",
                                    MF.J.a.58 = "GTGAATGTGAATAAGATCATCTTTGCTAAAGGAACGACACTGATTGTGAATGCTAGT",
                                    MF.J.a.59 = "GAATCTGGGAAATGACAAGCTGATGTTAGGATTTGGGACAAAACTAAACGTTTATTCCAGT",
                                    MF.J.a.60 = "GTGAATCAAGGAAATTACAAACTCATTATTGCTTCTGGGACACAATTAGTTGTTGATGCAAGT")
      
      
    }
    # Database beta
    { 
      CDR3.consensus.MF.b <- "SGIYY|NGLYL|SAVFY|TGVYL|IAVYF|SAVYF|SGVYF|AALYF|SGVYV"
      J.consensus.MF.b <- "FG[A-Z]GT[KR][LV]"
      MF.Vbs.4 = DNAStringSetList(
        MF.Vb_774708 = "ATGGCATGCATTCTTTTGAATGTGTTTGTGTTTCGCTTATGCCTTTCAGTTGATACTCTTGGGAAAACTGTTCTTCAGTCACCGGGAGATTTGATAAAAGAATCAAAAGCATCCACAGTCATTTTTTGCGCACACAGCATATCAAGCTATAACGTTATTATTTGGTACCGAGCAACTCGGAGCAGAAGCAATATGAAACTTATCGGATACATTTTATATGACAGCAAATATCCTGAACCTGATTTCAAAAGCAAAGTGGAATTGAGTGGAGATGGAAGAAGTGAAGCGTCTCTGAGCATCAGGAGTCTGGATCCATCAGACAGTGGCGTGTATTTCTGCGCAGCGTATGACACAGTGACTCACAATAAACCACGTGTGTCACAAAAACA",
        MF.Vb_776068= "ATGAACGGAAACATAATCCCACTCATCATTTGTGCACAATTATGGAAAGGTTGTGCGGGTGTGACGGAGGTTCTGCAGACTCCTGACGTTATATTTGGATTAATGAACACATCTGTCCAGATGCAGTGCACCCAGTATAAGGGCACAACACACAGGCTGATGTACTGGTACAAACAAAGGCCAGGGCAAAGCATCAGACTCCTGGCCTTCACAACCACTTACACCGAGGCAGAATACAGCAGCCGTGTGGAGCCGGACAAGCTGTGGGCCGTGAAGAACACTCCTGAGAGCGGATCTCTAACTGTGAGCCGCCTGGAGCCGGATGACGCCGCGCTTTACTTCTGTTCAGTCAGCGAGCACACGGTTCATCACGGCTGCTGGCAGACGCTACAAAAAGC",
        MF.Vb_777263= "ATGACTCACACTCTTGCTCTTGTCTCCCTAATTTGCCTTACAGTTGGGTCTCTGGGAAATACTGTTCTTCAGTCGCCTGGAGACTTGATAACAGAATCCAGCAAATCTGCAAGCATTCGTTGTTCACACAAAATACAAAGCTATGATCGTGTTCTCTGGTACAAACAAACTTTGGGCAAAACAAATGTGGAATTTATACTTCTTGGATATATTTTACTCGAACGCGAATATCCTGAACCTGATTTCAAAAGCAAAGTGGAATTGAGTGGAGATGGAAGAAGTGAGGCGTCTCTGAGCATCAGGAGTCTGGATCCATCAGACAGTGGCGTGTATTTCTGCGCAGCGTATGACACAGTGACTCACAATAAACCACGTGTGTCACAAAAACA",
        MF.Vb_779219= "ATGTTCACATTACTGCTCATCCTTTTGCTCTTTCCTTCAGATTTGGGTTTAAGGCAATCATCTGACATTTTAGTTCCGGAAGGTTTCAATGTGACTATAGTCTGTTCTCAAATTGGAACGTCGTTTAATGGCATGTACTGGTACAGGCACACAGCGGCCGGACTGGAGCTTGTGGTGTATTTCTATGTTAGACAAGCAAACTGGGAGAAAGACTTTGAGCAGAGAGCATCAGCATCCAGAAATGGAGCTTCTCTCAGCCTAACACTTGTAAAGGTGCAAAGTTCTGACAGCGCTGTTTATTTCTGCGCAAAGCAAGACGCACATAATAGTAGGAGCGTTTTTAACCTTAACAAAAACG",
        MF.Vb_780417 = "ATGATTAGTATCACAATTACTATCACTTTGATCCTTTTGTCATGCCTTTCTTACGTTGGTGGAGTGGACCAGTCTCCGGCTGATACATTTCAAAGACCTGGAGGTTCTGCAAGGATTGAGTGTTCACACAATATTCCCAATTACAACCAGATACTGTGGTACAAACAGACGCTGGGACACGAGAACGCTCCAATGATATTAATGGGATATTTATTATTGACGAATCCAACTAAAGAAAACAATTTTCAGAACCGAATAGAGATGTCTGGAAATGGAAATAGCAAAGGAGAGTTGGTCATCAAAGATCTCTCGCCTGACGACATTGCCGTTTATTTTTGTGCTGCGTGTTACACAGTACATAAGGTTCCAGTTTATTAATACAAAAATG",
        MF.Vb_782009 = "ATGATACAAAGTCCAGTTTTGTTTTTCCTGTTTTTTGTTTTGCAATGTGGTGAGTCGTTTGGCAACGTTTCCCAGTCTCCAGCGGATTTGATTAAAATGGCGGGACAGGAATTAATTCTGTCCTGCTCTCATGAACTCCCCGCCTATAACACCATACTATGGTATTATCTTCAAGATCAAAGCCTGGGATTTATTGGATATTCTTTTCACAGTTCACACACTTTGGCAAATGAAACTAAATATGAGATCAATGGAGATGGTAAAAGTTCAGTGACTCTAAGCATCTTCCAGCCGAGTCTGAGAGACACTGGTGTCTATCTCTGCGCTGCCAGTACACACTG",
        MF.Vb_782821 = "ATGCTCAACATGAAGGGAACTATTCTTCATACACTTACTATTATTCTCTTACTGGGAGTTGTGCTTTCTAGTAAAGTTGATCAATCACCTTCCTCCATTTTCTGGAGTCCTGATAATCCATCAGAAGTTGCCTGTCATCATCAAATTCCTAACTATGACACTATTCTCTGGTACCAGCAAACACAGCAAGGATCTGAGCTTAAACTCATCGGTAATGTGTACTATATGAGCGGCAAAGTAGAAACTTCATTTGAAAAATATTTTAACGTGAGTGGTAACGGTAAAACAAACTCTATTCTCCGTCTCCTCGTGAGACCATCAGAAGACAGCGCTGTGTTCTACTGTGCTGCTTATTACCACGCCGCTGTTTCCTCCCATCCTAATAAAACAAAAACC",
        MF.Vb_784246 = "ATGAACATGAGGGAGCTTCACATTTATCTGTTACTGTCTCTTTGTAGCACTGATCTGTCAGTCTCTCTAAGTGTTACGCAAACTCCAGTTAATCTTTTAGTGATGATGGGAGAATCCTTCAACCTCACATGCAGTCATCAAGACAAAAGCTATGACAAGATCTACTGGTACCATCAGCCCGAGCCTCAGACTCTCACACTGCTTGGTTTCCTCTCCTTTAAGCTGCCAGTGGATGTAAAGAACGACTACATGATTTCTGGCGATGCTGAGAAGGAGTCTGTCCTTCAGGTGTCTTCTGAAGGGGCTGAGCACAACGGATTATATCTCTGCGCTGTCAGTTCAGCACGGTGCTTGTAGTTCACACGGTTCTTCTCAAAAACCC",
        MF.Vb_784845 = "ATGACTTATTTAGCCACAATTATTATTTTCTCATTTCAAGCTTATGCTGTGTTTCCTCTTGACATCCATCAGTCTCCCAGAGAGCTTTTCCTGACAGTGGCGGAGAAAGAAACGGAGTTAATGTGCCAGCATGGAGACAGTAGCTACACTTACGTATACTGGTACCAGCAGATTGATGGTGGTGTAATGAATCTCATTGGAATGATCAATTATAAAATGCCTTCTGTAGAGAAAGATTTCAAATCTCGATTTAACATCACGGGAGAGGCTAAAAAAGAGGTATTTCTCTGAATCTCCAGCATTTCCACTAGTGACAGCGCAGTCTATTTCTGTGCATCAAGTTTGCACAGTACTGTATTTTCCCAGAGGTCCTAACAAAAAAAT",
        MF.Vb_785898 = "ATGTGGTTTTCCGAAAGCTATGTTTCTTGTGAAAAAATAATTTACTCCTTACTATTGTGTGGATGTCTGACAGACGGTGTTCAGTTTAATCAGCCGTCATCAATATTTATTACTCCAGGAGAAAAGGCAGAGATTTCCTGTTCTCACACGGATAGCAACTTGCTTGTAATGTTATGGTATCAACAACTCTATAACAGCACGTCTTTATCTCTCATCGGATATGGATATGGCACAAATGAGCCAGATGTTGAGTCTGTGTTCAAAGATCGATTTGAATTGACCAGGAAAAACACAGTGACGGGATCTTTACTTATCTCTAACCTCACTCTGCCTGACTCTGGTATCTACTATTGCGCTGCTAAAATACACGGCATTACAGTCTCAATGACTAGCCTGACTAAAAAC",
        MF.Vb_788076 = "ATGTGGTTTTCAAAAAGCCATGTTTCTTGTGAAGAAATCATTTACTCAATTCTTTTGTGTGCATGTCTGACAGACTGTGTTCAGTTTAATCAGCCGTTATCAATATTTATTACTCCAGGAGAAAAGGCAGAGATTTCCTGTTCTCACACGGATAGCAACCTAGATGTAATGTTATGGTATCAACAACTCTATAACAGCACGTCTTTGTCTCTCATCGGATATGGATATGCCACAACTGGCCCAAATTATGAGTCTGGGTTCAGTAGATTTAACATGACCAGGAAAAACACAGTGACGGGATCTTTACTTATCTCTAACCTCACTCTGCTTGACTCTGGTATCTACTATTGCGCTGCTAAAATACACGGCATTACAGTCTCAATGACTAGCCTGACTAAAAAC")
      
      MF.Vs.b.end = DNAStringSetList("TGCGCTGCTAAAATACACGGCATTACAGTCTCAATGACTAGCCTGACTAA",
                                     "TGCGCTGCTAAAATACACGGCATTACAGTCTCAATGACTAGCCTGACTAA",
                                     "TGTGCATCAAGTTTGCACAGTACTGTATTTTCCCAGAGGTCCTAACAAAA",
                                     "TGCGCTGTCAGTTCAGCACGGTGCTTGTAGTTCACACGGTTCTTCTCAAA",
                                     "TGTGCTGCTTATTACCACGCCGCTGTTTCCTCCCATCCTAATAAAACAAA",
                                     "TGCGCTGCCAGTACACACTGCAGCAGAACGCCTCCAGAGCTCACACAAGA",
                                     "TGTGCTGCGTGTTACACAGTACATAAGGTTCCAGTTTATTAATACAAAAA",
                                     "TGCGCAAAGCAAGACGCACATAATAGTAGGAGCGTTTTTAACCTTAACAA",
                                     "TGCGCAGCGTATGACACAGTGACTCACAATAAACCACGTGTGTCACAAAA",
                                     "TGTTCAGTCAGCGAGCACACGGTTCATCACGGCTGCTGGCAGACGCTACA",
                                     "TGCGCAGCGTATGACACAGTGACTCACAATAAACCACGTGTGTCACAAAA")
      
      MF.Vbs.last.36bp = subseq(unlist(MF.Vs.b.end),1,width=36)
      
      
      MF.Js.b = DNAStringSetList(MF.J.b.792035 = "TTTTACATTGTGACACTGGATCACAAGCGTATTTCGGGAACGGAACAAAACTAACTGTTTTAGGT",
                                 MF.J.b.792377 = "CAGTTTAAGGCTGTGCCAATTATCAAGCCTACTTTGGTAATGGAACAAAATTAACCGTTCTGGGT",
                                 MF.J.b.792604 = "TCATGCTGTGTTTCGGAAGCAGTCAAGCTTATTTTGGAAATGGAACAAAATTGACAGTTCTTGGT",
                                 MF.J.b.792774 = "TTGCCGCTTTGTGTTTAAACACGCAAGCAATATTCGGAAATGGGACAAAACTTACTGTTTTAGGT",
                                 MF.J.b.793098 = "TATATAATGTGTCAAGCACTGCACAAGCCTATTTTGGCAATGGAACAAAGTTAACTGTTCTAGGT",
                                 MF.J.b.793306 = "GAGTCAGTGTGAATAACTACAACCCTGCTTATTTTGGTGATGGAACCAAAGTTACCGTACTTGGT",
                                 MF.J.b.793638 = "GATGCTGTGACTCTGGTGGTTATCCTGCTCTTTTTGGCCATGGAACTAGGTTGTCCATTTTAGGT",
                                 MF.J.b.797892 = "AATCATCGTGCACTAGCTATGCTAAACTAGAATTTGGAGAAGGAACAAAACTGACAGTTCTGGGT")
      
      MF.Js.b.aligned = subseq(unlist(MF.Js.b),end = -10)
    }
    # Database gamma
    {
      CDR3.consensus.MF.g <- "AGWYY|AGTYY|SAVYY"
      
      J.consensus.MF.g <- "FG[A-Z]GT[RK]"
      
      MF.Vg.2 = DNAStringSetList(
        MF.Vg_1201219 = "ATGGAATTATATGGGATGCCAGCGTTTGTGCTTTCAGTTTTTTTTTTCTCTATAAATTGCGCACACGGCCAAATCAACCATCCTGTTCGGGCTATGATCAAACCCGCGGTTAAAACCGTTATAATTAAATGCTCGGTGCCCGAAGCGAGTCTTCAGTCCAGCCCGCTTCACCTGTACAGACGCACGGCAAATGGTTTTCAACGAATCATGCATTTTGCAACAGGGAATAAATTCGCCACAAACGACAACAACATTCCCAATAGATTTACCGGTTCAATAAAAAACAACGACGTGTCTCTCACCATCTCCTCGCTGAGTAAGGACGATGCTGGGTGGTATTACTGCGCGATGTGGAGCGGAGACACACATTGT",
        MF.Vg_1206250= "ATGTCTAGAATGGATAATGTGCTAACGTCTTTTATCTTCATCCTAATTAATTTCAAAGCAAATGGATCAGGGTTGCAACAGAAACTCTCCCTGACCAAAGGGGTGGACAAGTCGGCCAGTATAGACTGCTACATATATACAGACTGCTGGAGTTACGTCCACTGGTACCATCAAAAAGACTCAGAAGCTTTGAAGAGGATTTTGTATGCTGATCTGTTTGATGGGTCCACTAGAGAGACGAGTGCTGAGTCCATTAAATCAGAAATGAAAGCATCAAATTTATTTGTTTTAAAATTCACTAAGTTAAGGAAAGAGGACGCTGGCACGTATTACTGTGCTTGCTGGACCGAATACAGATATTTCACACAGTG",
        MF.Vg_1208519= "ATGAAGATGAGTCTTCATCTCCAGCTTTGTATTTCATTCTTTATTTTTTTTCAACGACCTGGAGCGTTGAGGCTTGAGCAGAAAATATCCATGACCAAAGCAGTGGACAAAACGGTGGTCCTAGACTGCAAATACCCCACAGACTGCAACTATTACATTCACTGGTATCAGCAGAAAGAAGGAGAAACCTTGAGGAGAATTCTGCATGCATCCATGAGCAATGGGGTAACATACAAGGATGCTGGCTTTGATGACTTCACAACAGTTAAGAAAAAGAGCAATTTAGCATTAAAAATTTCTGCAGTGAAAACTGAACATTCGGCCGTGTATTACTGCGCCTGCTGGTATTCTGCGTATTATAGTGGCACACAGCA"
      )
      
      MF_VGs.end = DNAStringSetList("TGCGCGATGTGGAGCGGAGACACACATTGTTCTCTCAGGCAGGAA",
                                    "TGTGCTTGCTGGACCGAATACAGATATTTCACACAGTGATCAAAA",
                                    "TGCGCCTGCTGGTATTCTGCGTATTATAGTGGCACACAGCAAGAC")
      
      MF.VGs.last.36bp = subseq(unlist(MF_VGs.end),end=-10)
      
      MF.Js.g = DNAStringSetList(
        MF.J.g.1 = "GCTGAAGAAAGAATGACTGTGGTAGACAAGAAAGTCTTCGGTTCGGGTACAAGACTTTATGT",
        MF.J.g.2 = "TTGGAATGAATCTCTGTGGGAAAGTGGATAAAAGAATTTGGTTCTGGAACAAAGCTGATAGT"
      )
      
      MF.Js.g.add = as.character(unlist(MF.Js.g))
    }
    # Database delta
    {
      CDR3.consensus.MF.d <- "SALYY|SAVYY|SAQYF|SAMYY|SGLYY|SALYF|AAVYY|SAKYY|SGLYF|INSKS|SVVYY|SALYH|SARYY"
      
      J.consensus.MF.d <- "FG[A-Z]PI[ST]L"
      
      MF.Js.d = DNAStringSetList(
        MF.J.d.1 = "GTCATGTGACTGAATGAAGTGGAGTCACCCCTAATCTTCGGAGATCCCATCACT",
        MF.J.d.2 = "GTTAGACGTGCATTTTCAACGGCCGACGCTTTAACGTTCGGAGAGCCGATATCC")
      
      MF.Js.d.add = as.character(unlist(MF.Js.d))
    }
    # Database IgH
    {
      CDR3.consensus.MF.h <- "TAVYY|MALYY|SAVYY|SATYY"
      J.consensus.MF.h <- "WG[A-Z]GT"
      
      MF.IgHs = DNAStringSetList(MF.IgH.Sc15.3169423 = "ATGACACATACACTTTTCTTTCTACTGTTCTTGCTCAATCTAAACCGTGTAAAATGTCAGAAGATGGAGTCGATAGAAAGCTCATTGCAGACGGAGCCTGGAAAGACTCTGACGTTGTCGTGTAGAGGATCTGGATTCAGCTTTGGCTGCTGTGACATGCACTGGATCAGGCAACAGGCCGGAAAGCCACTGGTGTGGATGGGACGTGTTTTAAGTGGTAGCAGTCATGACTATTCCGCCTCCTTTAAAGGCCGAATTGAGATCACCAGGGACAATTCCAACAGTTTGGTGTATCTGAAATTATCCCACTTGACTGAGGAGGATTCAGCCACTTATTACTGTGCAAAGCAAACACAGTG",
                                 MF.IgH.Sc15.3170621 = "ATGAATTCCACTTTCCAGTTGTGGCTGCTTCTAGCAGTAATCTCCTGCGCAGCTTGTGTTGAGCTGACACAGGCTGATTCTCTGGAGGTGAAACCTGGACAGACTTTGACTCTGACCTGTACAGTCTCTGGATATTCACTGACTGACAGCAGCTACTGTACTGACTGGATTAGACAACCCGCAGGAGAAGCTCTGGAATGGGTTGGTTGTATATGTGGTAGCGGTAACACCTACTACAGTGAAAAACTGAAAAGTAGATTACAAGTTTCCAGAGACTCTTCCAAAAGCACTGTGACGCTACAAAGCCAGAATATGAAGAGCGAAGACACGGCCGTTTATTACTGTGCCAGACAGACACCGTG",
                                 MF.IgH.Sc15.3171661 = "ATGACATTAATCATACACTCACTTCTGCTACTGCTTTTTATAACAGACACTTACAGTCAGACCTTCATTCAGTCTGATGCTGTGGTGAAAAGGCCAGATGAATCTTTTAAGCTTGTCTGCACTGCTTCTGGATTCACATTTGGTGACTACTGGGCAGCTTTAGTGAGACAAGCTCCAGGAAAAGGACTGGAATGGATTGCGTACATTAGCAGTGGCAGCGGCAGTACCACCTACTATTCCCAGTCTGTTCAGGGAAGATTCACAGTGTCCAGAGACAATTCCAGAAGTCAACTGTACCTTCAGATGAACAGACTCAAACCTGAAGACACGGCCGTGTATTACTGTGCCAGACAACCACAGTG",
                                 MF.IgH.Sc15.3173003 = "ATGGCAGGGTATTTATTATTGACTCTTTTCTTGTGCTTTGCCAAATTTCTAGGTACCAGTGGCCAGAGCCTGGAATCAATCCCTTTTAGCTCGGTCACCATAAAGCCTGGAGGAACTCTAAACATCTCCTGTAAAGGATCTGGCTTCAACTTTGGTCAGTATGGCATGCACTGGATCAGACAGCCGGCAGAGAAACCACTGAGCTGGATTGGAGCTATCTGGTACGATGCCAGTAAAACACTGTATGAAAATAGTGTAAAAGACCGAATAGAGATCACAAGAGACAACTCAAACAACATGGTGTATCTGAAACTCACTGATCTGTCGCCTGAGGACTCGGCTGTGTATTACTGTGCCAGATCAGCACAGTG",
                                 MF.IgH.Sc15.3173918 = "ATGACATTAATCATTCACTCACTTCTGCTACTGCTTTTTATAACAGATACTTACGGTCAGACCTTCATTCAGTCTGATGCTGTGGTGAAAAGGCCAGATGAATCTTTTAAGCTTGTCTGCACTGCTTCTGGATTCACATTTAGTGACTGTGCAGTTGTAGTGAGACAAGCTCCAGGAAAAGGACTGGAATGGATTGCGTACATTAGCAGTGGAAGCGGCAGTACCACCTACTATTCCCAGTCAGTTCAGGGAAGATTCACAGTGTCCAGAGACAATTCCAGGAGTCAACTGTACCTGCAGATGAACAGACTCAAACCTGAAGACATGGCCTTGTATTACTGTGCCAGAGAATCACAGTG",
                                 MF.IgH.Sc15.3175143 = "ATGTTCCTCATAAAGATCTTTCTGCTGTTGGCAGCTGTTCATGGAATCAACTCTCAGATTGTGCTGACACAGTCTGAACCGTCTCTGACTGCGTCTCCAAGTGGATCCCTGAAACTCACCTGTGCCTGCAGCGGATTCACTCCTAGCAGTTATTGGATGAGCTGGGTCCGACAGGCACCAGGGAAAGGACTGGAATGGATCATCAGCTATTATGATCCGGGCAATCCTACACACTCAGCTGCGTCAGTGCAGGGCAGATTCACAGCATCTAAAGACGCCACAAATTTCTATCTGCACATGAACCAGCTGAAGCCTGAAGACACGGCTGTGTATTACTGTGCTAGACAGCCACACTG",
                                 MF.IgH.Sc15.3176125 = "ATGAGAGTGTTGTGTCTCCTGCTGTTGGCCGCGTTGTCCTGTGTGGAGTCTCAGGTCGTGCTGACACAGTCTGAACAGTCAGTAGCTGTGGCTCCAGGAGGCTCCTACAAGCTCACCTGTGTCTGCAGTGGATCCTCACTTGGTAACATGTACTGGATTCGACAGGCACCGGGAAAAGGACTGGAATGGATCATTTATTATTATTCAGACACTGATAAGAGCTCAGCTTCGTCAGTGCAGGGCAGATTCACAGCATCTAAAGACGGCACAAATTTCTATCTGCACATGAACCAGCTGAAGCCTGAAGACACGGCTGTGTATTACTGTGCTAGAGAGCCACACTG",
                                 MF.IgH.Sc15.3177265 = "ATGGAGACGAGGATCATTATTCAGCTCATTATTTTAGGAATTTCAGCTCTTCAAGGAATCAACAGTGATGCAATTAGATTAGACCAGTCACCAGCTGTAATCAGGAAACCGGAGGAAACTATGAAGATCTCCTGCAAAATATCTGGTTTTACAATGACGAGCCATTATATGCACTGGATCAGACAGAAGGCAGGAAGACCTCTGGAGTGGATCGGGAGAGTGAACTGTGGCAGTAGCTCAGATACAATCTATGCGGACACCATGAAAAACCACTTCACTATGAGTGAAGACGTGTCTCTAGACACCCAGTACCTAGAAGCCAAGAGTCTGAGAGAAGAAGACACGGCTGTGTATTATTGTGCTCGAGAGACACAGTG",
                                 MF.IgH.Sc15.3178213 = "ATGGAAACGCTCCAGTTTGGACTATTTTTATTTTTTGTTTCATCGGTCAGAGCTCAGTCTCTCACATCTTCTAATGCGGTGATGACCAAACCTGGACAGTCCGTGACTCTGTCCTGCACTGTCTCTGGGTTTTCAGTGAGCAGCTACTACATGAACTGGATCCGCCAGAAACCAGGGAAAGGACTGGAGTGGATTGGCCGTATTGATGGTGGCACTGGCACTCGTTTTGCTCAGTTTGCACAAGACCAGTTCTCCATCACCAAAGACAACAGCTTAAACAAGTTGTATTTGGAGTTGAAAAGCCCAAAGACTGAAGACACAGCTGTTTATTATTGTGCAAGAGAGACACAGTG",
                                 MF.IgH.Sc15.3179277 = "ATGCCAGAAAAAATGCATTTCAGCTATGCCGTTGTTGTTTTTTTTCTATTTGTGATTCCAGATGTACAAGGTCAGTCACTCACTTCTTCTGGTCCTGAGATGAAGAAACCTGGGGATCCAGTTACATTGTCCTGTGTGGTATCTGGACTTTCTCTCAGCTGGTTGCACTGGATTCGTCAGAAACCAGGGAAAGGACTGGAGTGGATTGGCCGTATTGATAGTGGCACTGGCACTATTTTTGCCCAGTCTTTACAAGGCCAGTTTTCTATTACAAAAGACACAAGTAAAAACACGGTGCAACTGGAGGTGAAGAGCCTGAAGCCTGAAGACACAGCTGTGTATTACTGTGCACGAGAGTCACAGTA")
      
      MF_IgHs.end = DNAStringSetList("TGTGCAAAGCAAACACAGTGACACAAGAGTCTGGACCTG",
                                     "TGTGCCAGACAGACACCGTGATACAAACCACAGATGTGC",
                                     "TGTGCCAGACAACCACAGTGAATGAAAGTGAAGAAACAG",
                                     "TGTGCCAGATCAGCACAGTGTGTAAGATCTAGATTCAGC",
                                     "TGTGCCAGAGAATCACAGTGAATGAAAGTGAAGAAACAG",
                                     "TGTGCTAGACAGCCACACTGTTTCTGTTCATGCAGCTGC",
                                     "TGTGCTAGAGAGCCACACTGTTTCTGTTCTTACAGCTGC",
                                     "TGTGCTCGAGAGACACAGTGCAGCAGTTTACTGAAGCAC",
                                     "TGTGCAAGAGAGACACAGTGGACAAAAAGACATTGGAGC",
                                     "TGTGCACGAGAGTCACAGTAGCACAACAAGCTTTCTACT")
      
      MF.IgHs.last.36bp = subseq(unlist(MF_IgHs.end),end=-4)
      
      MF.Js.h = DNAStringSetList(MF.J.h.1 = "GAAGTGTTGGGAAATGTTTTTCGACAGTAAGTCAGGGGAATCAGTCACTGTGACTACTACTTTGACTACTGGGGGAAAGGAACCCAAGTGACTGTT",
                                 MF.J.h.2 = "CCAATTCGGACTGCGTTTTTACACACACGGCTCCATTAAATTTGTCAGTGTGACGGTGCCTTCGACTACTGGGGAAAGGGAACAATGGTGACAGTC",
                                 MF.J.h.3 = "TTATCAAAAAGCTTGTTTTTGTTGAGTGACTGAGGTGTAAAGTGTCACTGTGATTACGGTTTTGACTACTGGGGAAAAGGCACGTCTGTTACAGTC",
                                 MF.J.h.4 = "TGTATGGAATTGTTTTTTATATAACAATGTAAAGCAAATGCAGTAGTGTGTAGCTGGGCTTTAGATCACTGGGGCAGTGGCACCAAAGTCACCATA")
      
      MF.Js.h.add = subseq(as.character(unlist(MF.Js.h)),31,-6)
      
    }
    # Minifish alpha functions
    {
      MF.VAs.end.match = function (corrected.list){
        corrected.list$V.segment = match(corrected.list$V.segment,subseq(names(MF_Vas.5),1,8))
        V.length = length(MF.Vas.last.36bp)
        MF.Vas.last.36bp = append(MF.Vas.last.36bp,DNAStringSet(sub("TGTGCCCTG","TGCGCCCTG",MF.Vas.last.36bp[["Va478395"]])))
        MF.Vas.last.36bp = append(MF.Vas.last.36bp,DNAStringSet(sub("TGTGCACTG","TGCGCACTG",MF.Vas.last.36bp[["Va509325"]])))
        MF.Vas.last.36bp = append(MF.Vas.last.36bp,DNAStringSet(sub("TGCGCTCTG","TGCGCTCGG",MF.Vas.last.36bp[["Va527192"]])))
        MF.Vas.last.36bp = append(MF.Vas.last.36bp,DNAStringSet(sub("TGCGCTCTG","TGCGCCCTG",MF.Vas.last.36bp[["Va545980"]])))
        MF.Vas.last.36bp = append(MF.Vas.last.36bp,DNAStringSet(sub("TGCGCTCTG","TGCGCCCTG",MF.Vas.last.36bp[["Va571119"]])))
        MF.Vas.last.36bp = append(MF.Vas.last.36bp,DNAStringSet(sub("TGCGCTCTG","TGCGCCCTG",MF.Vas.last.36bp[["Va585213"]])))
        MF.Vas.last.36bp = append(MF.Vas.last.36bp,DNAStringSet(sub("TGCGCTCTG","TGCGCTCTT",MF.Vas.last.36bp[["Va589329"]])))
        names(MF.Vas.last.36bp) = 1:length(MF.Vas.last.36bp)
        
        
        MF.Vas.end.dict = list()
        MF.Vas.end.matrix.list = list()
        for (i in 1:36){
          MF.Vas.end.dict[[i]] = PDict(subseq(MF.Vas.last.36bp,end=-37+i))
          MF.Vas.end.matrix.list[[i]] = vcountPDict (MF.Vas.end.dict[[i]], subseq(DNAStringSet(corrected.list$CDR3) ,end=i))
        }
        
        MF.Vas.dataframe = data.frame("V.end"= numeric(length(corrected.list[,1])))
        
        for (i in 36:1){
          temp.which = which(MF.Vas.end.matrix.list[[i]]>0,arr.ind = TRUE) #reduce this expression to a variable to make it more ameanable
          temp.which[,1][temp.which[,1]==V.length+1] = grep("Va478395",names(MF_Vas.5)) #make the end of 29 be a 5 so the next line of code will recognize it as an allelle from the same V
          temp.which[,1][temp.which[,1]==V.length+2] =  grep("Va509325",names(MF_Vas.5)) 
          temp.which[,1][temp.which[,1]==V.length+3] =  grep("Va527192",names(MF_Vas.5)) 
          temp.which[,1][temp.which[,1]==V.length+4] =  grep("Va545980",names(MF_Vas.5)) 
          temp.which[,1][temp.which[,1]==V.length+5] =  grep("Va571119",names(MF_Vas.5)) 
          temp.which[,1][temp.which[,1]==V.length+6] =  grep("Va585213",names(MF_Vas.5)) 
          temp.which[,1][temp.which[,1]==V.length+7] =  grep("Va589329",names(MF_Vas.5)) 
          
          length.i.matches <- temp.which[,2][corrected.list$V.segment[temp.which[,2]]==temp.which[,1]] #finds the matching V entries that also match V segment. Doublechecks V (so to speak, at least if there is no match the result will be a very short V before the nontemplate nucleotides)
          MF.Vas.dataframe$V.end[length.i.matches[MF.Vas.dataframe$V.end[length.i.matches]==0]] = i
        }
        return(MF.Vas.dataframe)
      }
      
      MF.Js.a.dataframe.creator = function (clone.list){
        MF.Js.dict.list = list()
        MF.Js.matrix.list = list()
        
        for (i in 1:50){
          MF.Js.dict.list[[i]] =  PDict(subseq(MF.Js.a.aligned,start=i))
          MF.Js.matrix.list[[i]] <- vcountPDict (MF.Js.dict.list[[i]], DNAStringSet(clone.list$CDR3 ))
        }
        
        MF.Js.dataframe = data.frame("J segment" = numeric(nrow(clone.list)), "J start" = numeric(nrow(clone.list)))
        
        for (i in 1:50){
          MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]]==0] = which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,1] [MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0]
          MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0] = i
        }
        return (MF.Js.dataframe)
      }
      
      MF.Ja_call = function(clone.list){
        short.CDR3s = clone.list[nchar(clone.list)<69]
        correctivo = nchar(short.CDR3s)
        correctivo2 = nchar(clone.list)<69
        #I prolongue the shorter CDR3s to be able to perform a match search, with a DIFFERENT character than I extended Js "X" vs "N"
        X.CDR3s = unlist(lapply(short.CDR3s,function(x)paste0(paste0(c(rep("X",(69-nchar(x)))),collapse = ""),x,collapse = ""))) 
        #And replace
        clone.list[nchar(clone.list)<69] = X.CDR3s
        
        J.matches = lapply(21:69,function(i)match(subseq(clone.list,-i,-15),subseq(MF.Js.a.aligned,(73-i)),nomatch = 0))
        
        J.start = rep(0,length(clone.list))
        J.matches.final = rep(0,length(clone.list))
        for (k in length(J.matches):1){
          J.start[J.start == 0][J.matches[[k]][J.start == 0] != 0 ] = k
          J.matches.final[J.matches.final == 0] = J.matches[[k]][J.matches.final==0]
        }
        #J.start[correctivo2] = J.start[correctivo2]-(69-correctivo)
        
        #correction for ambiguous calling
        sD.Ja = lapply(21:69,function(i)stringDist(subseq(MF.Js.a.aligned,(73-i)),method = "hamming"))
        conflicting.clusters = lapply(1:49,function(i)graph.builder(which(matrix.builder(sD.Ja[[i]])==0,arr.ind = TRUE)))
        df.Ja = data.frame(J.matches.final,J.start)
        for (i in 1:length(conflicting.clusters)){
          if (length(conflicting.clusters[[i]])>0)
            for (j in 1:length(conflicting.clusters[[i]])){
              df.Ja$J.matches.final[df.Ja$J.start == i][df.Ja[df.Ja$J.start == i,]$J.matches.final %in% as.numeric(conflicting.clusters[[i]][[j]])] = paste0(as.numeric(conflicting.clusters[[i]][[j]]),collapse = ",")
            }
        }
        
        return(df.Ja)
      }
      
      MF.TA.data.parser = function(corrected.clone.list){
        df.final = corrected.clone.list
        groupColumns = c("CDR3","V.segment")
        dataColumns = c("Freq")
        df.final = ddply(df.final, groupColumns, list(UMI.count = nrow, Total = function(x) colSums(x[dataColumns])))
        
        #Make Vs end list
        df.final$V.end = MF.VAs.end.match(df.final)
        
        # Make J list
        Js.df = MF.Js.a.dataframe.creator(df.final)
        Ja.df = MF.Ja_call(df.final$CDR3)
        
        df.final$J.segment = Js.df$J.segment
        df.final$J.start = nchar(as.character(df.final$CDR3))+Js.df$J.start-72
        
        #add a few more columns
        df.final$CDR3.C.to.F = as.character(subseq(DNAStringSet(df.final$CDR3,1,-28)))
        df.final$CDR3.aa= as.character(translate(subseq(DNAStringSet(df.final$CDR3,1,-28))))
        
        
        df.umi.sum = sum(df.final$UMI.count)
        df.reads.sum = sum(df.final$Total.Freq)
        
        df.for.tcR = data.frame("Umi.count" =df.final$UMI.count , "Umi.proportion" =df.final$UMI.count/df.umi.sum , "Read.count" = df.final$Total.Freq , "Read.proportion" =  df.final$Total.Freq/df.reads.sum , "CDR3.nucleotide.sequence" = df.final$CDR3.C.to.F , "CDR3.amino.acid.sequence" = df.final$CDR3.aa , "V.gene" = as.character(df.final$V.segment) , "J.gene" = as.character(df.final$J.segment), "D.gene" = rep("", length(df.final$UMI)) , "V.end" = df.final$V.end , "J.start" = df.final$J.start, "D5.end" = rep(-1, length(df.final$UMI)), "D3.end" = rep(-1, length(df.final$UMI)) , "VD.insertions" = rep(-1, length(df.final$UMI)) , "DJ.insertions" = rep(-1, length(df.final$UMI)) , "Total.insertions" = as.numeric(unlist(df.final$J.start -df.final$V.end -1)) ,stringsAsFactors = FALSE)  
        return(df.for.tcR)
      }
      
    }
    # Minifish beta functions
    {
      MF.VBs.end.match = function (corrected.list){
        V.length = length(MF.Vbs.last.36bp)
        MF.Vbs.last.36bp = append(MF.Vbs.last.36bp,DNAStringSet(sub("TGCGCTGTCAG", "TGTGCTGTCAG",MF.Vbs.last.36bp[3])))
        MF.Vbs.last.36bp = append(MF.Vbs.last.36bp,DNAStringSet(sub("TGCGCTGC","TGTGCTGC",MF.Vbs.last.36bp[5])))
        ### First makes [V.length] dictionaries of length 22 to 1 and match to subsets of the sequences starting from left, same length
        
        MF.Vbs.end.dict = list()
        MF.Vbs.end.matrix.list = list()
        for (i in 1:36){
          MF.Vbs.end.dict[[i]] = PDict(subseq(MF.Vbs.last.36bp,end=-37+i))
          MF.Vbs.end.matrix.list[[i]] = vcountPDict (MF.Vbs.end.dict[[i]], subseq(DNAStringSet(corrected.list$CDR3) ,end=i))
        }
        
        MF.Vbs.dataframe = data.frame("V.end"= numeric(length(corrected.list[,1])))
        
        for (i in 36:1){
          temp.which = which(MF.Vbs.end.matrix.list[[i]]>0,arr.ind = TRUE) #reduce this expression to a variable to make it more ameanable
          temp.which[,1][temp.which[,1]==V.length+1] = 3
          temp.which[,1][temp.which[,1]==V.length+1] = 5
          
          length.i.matches <- temp.which[,2][corrected.list$V.segment[temp.which[,2]]==names(MF.Vbs.4)[temp.which[,1]]] #finds the matching V entries that also match V segment. Doublechecks V (so to speak, at least if there is no match the result will be a very short V before the nontemplate nucleotides)
          MF.Vbs.dataframe$V.end[length.i.matches[MF.Vbs.dataframe$V.end[length.i.matches]==0]] = i
        }
        return(MF.Vbs.dataframe)
      }
      
      MF.Js.b.dataframe.creator = function (clone.list){
        MF.Js.dict.list = list()
        MF.Js.matrix.list = list()
        
        for (i in 1:40){
          MF.Js.dict.list[[i]] =  PDict(subseq(MF.Js.b.aligned,start=i))
          MF.Js.matrix.list[[i]] <- vcountPDict (MF.Js.dict.list[[i]], DNAStringSet(clone.list$CDR3 ))
        }
        
        MF.Js.dataframe = data.frame("J segment" = numeric(nrow(clone.list)), "J start" = numeric(nrow(clone.list)))
        
        for (i in 1:40){
          MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]]==0] = which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,1] [MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0]
          MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0] = i
        }
        return (MF.Js.dataframe)
      }
      
      MF.TB.data.parser.v2 = function(corrected.clone.list){
        df.final = corrected.clone.list
        groupColumns = c("CDR3","V.segment")
        dataColumns = c("Freq")
        df.final = ddply(df.final, groupColumns, list(UMI.count = nrow, Total = function(x) colSums(x[dataColumns])))
        
        #Make Vs end list
        df.final$V.end = MF.VBs.end.match(df.final)
        
        # Make J list
        Js.df = MF.Js.b.dataframe.creator(df.final)
        df.final$J.segment = Js.df$J.segment
        df.final$J.start = nchar(as.character(df.final$CDR3))+Js.df$J.start-62
        
        #add a few more columns
        df.final$CDR3.C.to.F = as.character(subseq(DNAStringSet(df.final$CDR3,1,-28)))
        df.final$CDR3.aa= as.character(translate(subseq(DNAStringSet(df.final$CDR3,1,-28))))
        
        df.umi.sum = sum(df.final$UMI.count)
        df.Freq.sum = sum(df.final$Total.Freq)
        Db = "GTGGGGACAGGGGGCCACGCCACGGTG"
        
        Db.kmers = sapply(1:nchar(Db),function(i)kmer.generator(Db,largo = i))
        Db.matches = sapply(nchar(Db):1,function(i)sapply(1:length(Db.kmers[[i]]),function(j)regexpr(Db.kmers[[i]][j],df.final$CDR3.C.to.F)))
        Db.indexes = lapply(Db.matches,function(x)apply(x,1,function(y)if(sum(y!= -1) ==1) y[y!= -1]else 0))
        names(Db.indexes) = nchar(Db):1
        
        D.start = rep(-1,length(df.final$CDR3.C.to.F))
        D.length = D.start
        D.end = D.start
        for (i in 1:(length(Db.indexes)-5))D.start[which(Db.indexes[[i]]!=0)] = Db.indexes[[i]][which(Db.indexes[[i]]!=0)]
        for (i in 1:(length(Db.indexes)-5))D.length[which(Db.indexes[[i]]!=0)] = as.numeric(names(Db.indexes[i]))
        D.end = D.length+D.start-1
        D.end[D.end == -3] = -1
        VD.i = D.start-as.numeric(unlist(df.final$V.end))-1 
        VD.i[D.start== -1] = NA
        barplot(table(VD.i))
        DJ.i = as.numeric(df.final$J.start) - D.end -1
        DJ.i[D.end== -1] = NA
        df.for.tcR = data.frame("Umi.count" =df.final$UMI.count , "Umi.proportion" =df.final$UMI.count/df.umi.sum , "Read.count" = df.final$Total.Freq , "Read.proportion" =  df.final$Total.Freq/df.Freq.sum , "CDR3.nucleotide.sequence" = df.final$CDR3.C.to.F , "CDR3.amino.acid.sequence" = df.final$CDR3.aa , "V.gene" = as.character(df.final$V.segment) , "J.gene" = as.character(df.final$J.segment), "D.gene" = rep("", length(df.final$UMI)) , "V.end" = df.final$V.end , "J.start" = df.final$J.start, "D5.end" = D.start, "D3.end" = D.end , "VD.insertions" = VD.i , "DJ.insertions" = DJ.i , "Total.insertions" = as.numeric(unlist(df.final$J.start -df.final$V.end -1)) ,stringsAsFactors = FALSE)  
        return(df.for.tcR)
      }
    }
    # Minifish gamma functions
    {
      Canonical.CDR3.j.trimmer.MF.TG.Mk1 <- function(filtered.translated.reads,filtered.r1.reads,reading.frames.table){
        #Mk_IV: changed CDR3 consensus to ".*[AGST][VILEMTG][FYW][FYLSI]C" (added T in first position)
        cat("Canonical.CDR3.j.trimmer.MDK.Mk_V ","\n",sep = "")
        inicio=Sys.time()
        
        
        J.match = regexpr(J.consensus,filtered.translated.reads)
        C.match.rf.1 = regexpr(CDR3.consensus.1,filtered.translated.reads[J.match!=-1])
        C.match.rf.2 = regexpr(CDR3.consensus.2,filtered.translated.reads[J.match!=-1][C.match.rf.1==-1])
        C.match.rf.3 = regexpr(CDR3.consensus.3,filtered.translated.reads[J.match!=-1][C.match.rf.1==-1][C.match.rf.2==-1])
        ### Aminoacid sequences
        filtro.rf1=(1:length(filtered.translated.reads))[J.match!=-1][C.match.rf.1!=-1]
        filtro.rf2=(1:length(filtered.translated.reads))[J.match!=-1][C.match.rf.1==-1][C.match.rf.2!=-1]
        filtro.rf3=(1:length(filtered.translated.reads))[J.match!=-1][C.match.rf.1==-1][C.match.rf.2==-1][C.match.rf.3!=-1]
        principio.1=(C.match.rf.1[C.match.rf.1!=-1]+5)*3-3+reading.frames.table[filtro.rf1]
        largo.1=nchar(subseq(filtered.translated.reads[filtro.rf1],C.match.rf.1[C.match.rf.1!=-1]+5,J.match[filtro.rf1]))
        nucleotide.CDR3.1 <- subseq(filtered.r1.reads[filtro.rf1],start=principio.1,width = largo.1*3+27)
        principio.2=(C.match.rf.2[C.match.rf.2!=-1]+5)*3-2+reading.frames.table[filtro.rf2]
        largo.2=nchar(subseq(filtered.translated.reads[filtro.rf2],C.match.rf.2[C.match.rf.2!=-1]+4,J.match[filtro.rf2]))
        largo.2[largo.2<=1]=1
        nucleotide.CDR3.2 <- subseq(filtered.r1.reads[filtro.rf2],start=principio.2,width = (largo.2-1)*3+26)
        principio.3=(C.match.rf.3[C.match.rf.3!=-1]+5)*3-1+reading.frames.table[filtro.rf3]
        largo.3=nchar(subseq(filtered.translated.reads[filtro.rf3],C.match.rf.3[C.match.rf.3!=-1]+4,J.match[filtro.rf3]))
        largo.3[largo.3<=1]=2
        nucleotide.CDR3.3 <- subseq(filtered.r1.reads[filtro.rf3],start=principio.3,width = (largo.3-1)*3+25)
        
        nucleotide.CDR3.list <- list(nucleotide.CDR3.1, nucleotide.CDR3.2, nucleotide.CDR3.3)
        canonical.filter.list <-  list(filtro.rf1, filtro.rf2, filtro.rf3)
        
        print(round(difftime(Sys.time(),inicio,units = "auto"),digits = 2))
        return(list(CDR3s=nucleotide.CDR3.list,CDR3.filter=canonical.filter.list))
      }
      
      MF.VGs.end.match=function (corrected.list){
        V.length = length(MF.VGs.last.36bp)
        
        MF.VGs.last.36bp = append(MF.VGs.last.36bp,DNAStringSet(sub("TGTGCTTGCTGGACCGAATACAGATATTTCACACAG","TGTGCTTGCTGGGCCTGGGATTCCAGATATTTCACA",MF.VGs.last.36bp[2]))) #found in female assembly
        MF.VGs.last.36bp = append(MF.VGs.last.36bp,DNAStringSet(sub("TGCGCCTGCTGGTATTCTGCGTATTATAGTGGCAC","TGCGCCTGCTGGGATTATAGTGTCACACAGCAAGA",MF.VGs.last.36bp[3]))) #found in female assembly

        MF.VGs.end.dict = list()
        MF.VGs.end.matrix.list = list()
        for (i in 1:36){
          MF.VGs.end.dict[[i]] = PDict(subseq(MF.VGs.last.36bp,end=-37+i))
          MF.VGs.end.matrix.list[[i]] = vcountPDict (MF.VGs.end.dict[[i]], subseq(DNAStringSet(corrected.list$CDR3) ,end=i))
        }
        
        MF.VGs.dataframe = data.frame("V.end"= numeric(length(corrected.list[,1])))
        
        for (i in 36:1){
          temp.which = which(MF.VGs.end.matrix.list[[i]]>0,arr.ind = TRUE) #reduce this expression to a variable to make it more ameanable
          temp.which[,1][temp.which[,1]==V.length+1] = 2
          temp.which[,1][temp.which[,1]==V.length+2] = 3
          temp.which[,1][temp.which[,1]==V.length+3] = 3
          temp.which[,1][temp.which[,1]==V.length+4] = 3
          length.i.matches <- temp.which[,2][match(corrected.list$V.segment[temp.which[,2]],names(MF.Vg.2))==temp.which[,1]] 
          MF.VGs.dataframe$V.end[length.i.matches[MF.VGs.dataframe$V.end[length.i.matches]==0]] = i
        }
        return(MF.VGs.dataframe)
      }
      
      
      MF.Js.g.dataframe.creator = function (clone.list){
        MF.Js.dict.list = list()
        MF.Js.matrix.list = list()
        
        for (i in 1:50){
          MF.Js.dict.list[[i]] =  PDict(subseq(MF.Js.g.add,start=i))
          MF.Js.matrix.list[[i]] <- vcountPDict (MF.Js.dict.list[[i]], DNAStringSet(clone.list$CDR3 ))
        }
        
        MF.Js.dataframe = data.frame("J segment" = numeric(nrow(clone.list)), "J start" = numeric(nrow(clone.list)))
        
        for (i in 1:50){
          MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]]==0] = which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,1] [MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0]
          MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0] = i
        }
        return (MF.Js.dataframe)
      }
      
      MF.TG.data.parser = function(corrected.clone.list){
        df.final = corrected.clone.list
        groupColumns = c("CDR3","V.segment")
        dataColumns = c("Freq")
        df.final = ddply(df.final, groupColumns, list(UMI.count = nrow, Total = function(x) colSums(x[dataColumns])))
        
        #Make Vs end list
        df.final$V.end = MF.VGs.end.match(df.final)
        
        # Make J list
        Js.df = MF.Js.g.dataframe.creator(df.final)
        df.final$J.segment = Js.df$J.segment
        df.final$J.start = nchar(as.character(df.final$CDR3))+Js.df$J.start-66
        
        #add a few more columns
        df.final$CDR3.C.to.F = as.character(subseq(DNAStringSet(df.final$CDR3,1,-28)))
        df.final$CDR3.aa= as.character(translate(subseq(DNAStringSet(df.final$CDR3,1,-28))))
        
        df.umi.sum = sum(df.final$UMI.count)
        df.Freq.sum = sum(df.final$Total.Freq)
        
        df.for.tcR = data.frame("Umi.count" =df.final$UMI.count , "Umi.proportion" =df.final$UMI.count/df.umi.sum , "Read.count" = df.final$Total.Freq , "Read.proportion" =  df.final$Total.Freq/df.Freq.sum , "CDR3.nucleotide.sequence" = df.final$CDR3.C.to.F , "CDR3.amino.acid.sequence" = df.final$CDR3.aa , "V.gene" = as.character(df.final$V.segment) , "J.gene" = as.character(df.final$J.segment), "D.gene" = rep("", length(df.final$UMI)) , "V.end" = df.final$V.end , "J.start" = df.final$J.start, "D5.end" = rep(-1, length(df.final$UMI)), "D3.end" = rep(-1, length(df.final$UMI)) , "VD.insertions" = rep(-1, length(df.final$UMI)) , "DJ.insertions" = rep(-1, length(df.final$UMI)) , "Total.insertions" = as.numeric(unlist(df.final$J.start -df.final$V.end -1)) ,stringsAsFactors = FALSE)  
        return(df.for.tcR)
      }
    }
    # Minifish delta functions
    {
      MF.Js.d.dataframe.creator = function (clone.list){
        MF.Js.dict.list = list()
        MF.Js.matrix.list = list()
        
        for (i in 1:50){
          MF.Js.dict.list[[i]] =  PDict(subseq(MF.Js.d.add,start=i))
          MF.Js.matrix.list[[i]] <- vcountPDict (MF.Js.dict.list[[i]], DNAStringSet(clone.list$CDR3 ))
        }
        
        MF.Js.dataframe = data.frame("J segment" = numeric(nrow(clone.list)), "J start" = numeric(nrow(clone.list)))
        
        for (i in 1:50){
          MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,2]]==0] = which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,1] [MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0]
          MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0] = i
        }
        return (MF.Js.dataframe)
      }
      
      MF.TD.data.parser = function(corrected.clone.list){
        df.final = corrected.clone.list
        groupColumns = c("CDR3","V.segment")
        dataColumns = c("Freq")
        df.final = ddply(df.final, groupColumns, list(UMI.count = nrow, Total = function(x) colSums(x[dataColumns])))
        
        #Make Vs end list
        df.final$V.end = MF.VAs.end.match(df.final)
        
        # Make J list
        Js.df = MF.Js.d.dataframe.creator(df.final)
        df.final$J.segment = Js.df$J.segment
        df.final$J.start = nchar(as.character(df.final$CDR3))+Js.df$J.start-66
        
        
        #add a few more columns
        df.final$CDR3.C.to.F = as.character(subseq(DNAStringSet(df.final$CDR3,1,-28)))
        df.final$CDR3.aa= as.character(translate(subseq(DNAStringSet(df.final$CDR3,1,-28))))
        
        
        
        
        df.umi.sum = sum(df.final$UMI.count)
        df.Freq.sum = sum(df.final$Total.Freq)
        
        Dd1 = "GTGGATTGGGGTACCACAGTG" #428281
        Dd2 = "GTGGATACGTTATTACCACAGTG" #428545
        
        
        Dd1.kmers = sapply(1:nchar(Dd1),function(i)kmer.generator(Dd1,largo = i))
        Dd2.kmers = sapply(1:nchar(Dd2),function(i)kmer.generator(Dd2,largo = i))
        Dd1.matches = sapply(nchar(Dd1):1,function(i)sapply(1:length(Dd1.kmers[[i]]),function(j)regexpr(Dd1.kmers[[i]][j],df.final$CDR3.C.to.F)))
        Dd2.matches = sapply(nchar(Dd2):1,function(i)sapply(1:length(Dd2.kmers[[i]]),function(j)regexpr(Dd2.kmers[[i]][j],df.final$CDR3.C.to.F)))
        
        Dd1.indexes = lapply(Dd1.matches,function(x)apply(x,1,function(y)if(sum(y!= -1) ==1) y[y!= -1]else 0))
        Dd2.indexes = lapply(Dd2.matches,function(x)apply(x,1,function(y)if(sum(y!= -1) ==1) y[y!= -1]else 0))
        names(Dd1.indexes) = nchar(Dd1):1
        names(Dd2.indexes) = nchar(Dd2):1
        
        
        D1.start = rep(-1,length(df.final$CDR3.C.to.F))
        D1.length = D1.start
        D1.end = D1.start
        for (i in 1:(length(Dd1.indexes)-6))D1.start[which(Dd1.indexes[[i]]!=0)] = Dd1.indexes[[i]][which(Dd1.indexes[[i]]!=0)]
        for (i in 1:(length(Dd1.indexes)-6))D1.length[which(Dd1.indexes[[i]]!=0)] = as.numeric(names(Dd1.indexes[i]))
        D1.end = D1.length+D1.start-1
        D1.end[D1.end == -3] = -1
        
        D2.start = rep(-1,length(df.final$CDR3.C.to.F))
        D2.length = D2.start
        D2.end = D2.start
        for (i in 1:(length(Dd2.indexes)-6))D2.start[which(Dd2.indexes[[i]]!=0)] = Dd2.indexes[[i]][which(Dd2.indexes[[i]]!=0)]
        for (i in 1:(length(Dd2.indexes)-6))D2.length[which(Dd2.indexes[[i]]!=0)] = as.numeric(names(Dd2.indexes[i]))
        D2.end = D2.length+D2.start-1
        D2.end[D2.end == -3] = -1
        
        VD.i1 = D1.start-as.numeric(unlist(df.final$V.end))-1 
        VD.i1[D1.start== -1] = NA
        #barplot(table(VD.i1))
        DJ.i1 = as.numeric(df.final$J.start) - D1.end -1
        DJ.i1[D1.end== -1] = NA
        
        VD.i2 = D2.start-as.numeric(unlist(df.final$V.end))-1 
        VD.i2[D2.start== -1] = NA
        #barplot(table(VD.i2))
        DJ.i2 = as.numeric(df.final$J.start) - D2.end -1
        DJ.i2[D2.end== -1] = NA
        DD.insertions = D2.start-D1.end-1
        DD.insertions[D2.start == -1 | D1.end == -1] = NA
        df.for.tcR = data.frame("Umi.count" =df.final$UMI.count , "Umi.proportion" =df.final$UMI.count/df.umi.sum , "Read.count" = df.final$Total.Freq , "Read.proportion" =  df.final$Total.Freq/df.Freq.sum , "CDR3.nucleotide.sequence" = df.final$CDR3.C.to.F , "CDR3.amino.acid.sequence" = df.final$CDR3.aa , "V.gene" = as.character(df.final$V.segment) , "J.gene" = as.character(df.final$J.segment), "D.gene" = rep("", length(df.final$UMI)) , "V.end" = df.final$V.end , "J.start" = df.final$J.start, "D5_1.end" =  D1.start, "D1_3.end" =  D1.end ,"D2_5.end" =  D2.start, "D2_3.end" =  D2.end , "VD.insertions" = VD.i1 , "DJ.insertions" = DJ.i1 , "Total.insertions" = as.numeric(unlist(df.final$J.start -df.final$V.end -1)) ,stringsAsFactors = FALSE)  
        return(df.for.tcR)
      }
    }
    # Minifish IgH functions
    {
      MF.IgHs.end.match = function (corrected.list){
        V.length = length(MF.IgHs.last.36bp)
        corrected.list$V.segment = match(corrected.list$V.segment,names(MF.IgHs))
        MF.IgHs.last.36bp = append(MF.IgHs.last.36bp,DNAStringSet(sub("TGTGCCAGA","AGTGCCAGA",MF.IgHs.last.36bp[2])))
        names(MF.IgHs.last.36bp) = 1:length(MF.IgHs.last.36bp)
        
        ### First makes [V.length] dictionaries of length 22 to 1 and match to subsets of the sequences starting from left, same length
        
        MF.IgHs.end.dict = list()
        MF.IgHs.end.matrix.list = list()
        for (i in 1:35){
          MF.IgHs.end.dict[[i]] = PDict(subseq(MF.IgHs.last.36bp,end=-37+i))
          MF.IgHs.end.matrix.list[[i]] = vcountPDict (MF.IgHs.end.dict[[i]], subseq(DNAStringSet(corrected.list$CDR3) ,end=i))
        }
        
        MF.IgHs.dataframe = data.frame("V.end"= numeric(length(corrected.list[,1])))
        
        for (i in 35:1){
          temp.which = which(MF.IgHs.end.matrix.list[[i]]>0,arr.ind = TRUE) #reduce this expression to a variable to make it more ameanable
          temp.which[,1][temp.which[,1]==V.length+1] = grep("MF.IgH.Sc15.3170621",names(MF.IgHs))
          length.i.matches <- temp.which[,2][corrected.list$V.segment[temp.which[,2]]==temp.which[,1]] #finds the matching V entries that also match V segment. Doublechecks V (so to speak, at least if there is no match the result will be a very short V before the nontemplate nucleotides)
          MF.IgHs.dataframe$V.end[length.i.matches[MF.IgHs.dataframe$V.end[length.i.matches]==0]] = i
        }
        return(MF.IgHs.dataframe)
      }
      
      MF.Js.h.dataframe.creator = function (clone.list){
        #should in principle handle 2 Js, 
        MF.Js.dict.list = list()
        MF.Js.matrix.list = list()
        
        for (i in 1:50){
          MF.Js.dict.list[[i]] =  PDict(subseq(MF.Js.h.add,start=i))
          MF.Js.matrix.list[[i]] <- vcountPDict (MF.Js.dict.list[[i]], DNAStringSet(clone.list$CDR3 ))
        }
        
        MF.Js.dataframe = data.frame("J segment" = numeric(nrow(clone.list)), "J start" = numeric(nrow(clone.list)))
        
        for (i in 1:50){
          MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0] = which(MF.Js.matrix.list[[i]]==1,arr.ind = TRUE)[,1] [MF.Js.dataframe$J.segment[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0]
          MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]][MF.Js.dataframe$J.start[which(MF.Js.matrix.list[[i]]>0,arr.ind = TRUE)[,2]]==0] = i
        }
        return (MF.Js.dataframe)
      }
      
      MF.IgH.data.parser = function(corrected.clone.list){
        
        df.final = corrected.clone.list
        groupColumns = c("CDR3","V.segment")
        dataColumns = c("Freq")
        df.final = ddply(df.final, groupColumns, list(UMI.count = nrow, Total = function(x) colSums(x[dataColumns])))
        
        #Make Vs end list
        df.final$V.end = MF.IgHs.end.match(df.final)
        
        # Make J list
        Js.df = MF.Js.h.dataframe.creator(df.final)
        df.final$J.segment = Js.df$J.segment
        df.final$J.start = nchar(as.character(df.final$CDR3))+Js.df$J.start-69
        
        
        #add a few more columns
        df.final$CDR3.C.to.F = as.character(subseq(DNAStringSet(df.final$CDR3,1,-28)))
        df.final$CDR3.aa= as.character(translate(subseq(DNAStringSet(df.final$CDR3,1,-28))))
        
        df.umi.sum = sum(df.final$UMI.count)
        df.Freq.sum = sum(df.final$Total.Freq)
        
        df.for.tcR = data.frame("Umi.count" =df.final$UMI.count , "Umi.proportion" =df.final$UMI.count/df.umi.sum , "Read.count" = df.final$Total.Freq , "Read.proportion" =  df.final$Total.Freq/df.Freq.sum , "CDR3.nucleotide.sequence" = df.final$CDR3.C.to.F , "CDR3.amino.acid.sequence" = df.final$CDR3.aa , "V.gene" = as.character(df.final$V.segment) , "J.gene" = as.character(df.final$J.segment), "D.gene" = rep("", length(df.final$UMI)) , "V.end" = df.final$V.end , "J.start" = df.final$J.start, "D5.end" = rep(-1, length(df.final$UMI)), "D3.end" = rep(-1, length(df.final$UMI)) , "VD.insertions" = rep(-1, length(df.final$UMI)) , "DJ.insertions" = rep(-1, length(df.final$UMI)) , "Total.insertions" = as.numeric(unlist(df.final$J.start -df.final$V.end -1)) ,stringsAsFactors = FALSE)  
        return(df.for.tcR)
      }
      
      MF.IgH.data.parser = function(corrected.clone.list){
        
        df.final = corrected.clone.list
        groupColumns = c("CDR3","V.segment")
        dataColumns = c("Freq")
        df.final = ddply(df.final, groupColumns, list(UMI.count = nrow, Total = function(x) colSums(x[dataColumns])))
        
        #Make Vs end list
        df.final$V.end = MF.IgHs.end.match(df.final)
        
        # Make J list
        Js.df = MF.Js.h.dataframe.creator(df.final)
        df.final$J.segment = Js.df$J.segment
        df.final$J.start = nchar(as.character(df.final$CDR3))+Js.df$J.start-69
        
        
        #add a few more columns
        df.final$CDR3.C.to.F = as.character(subseq(DNAStringSet(df.final$CDR3,1,-28)))
        df.final$CDR3.aa= as.character(translate(subseq(DNAStringSet(df.final$CDR3,1,-28))))
        
        df.umi.sum = sum(df.final$UMI.count)
        df.Freq.sum = sum(df.final$Total.Freq)
        
        Dh1 = "CACTGTGTATGGTGGGGCGGCACAGTGACA"
        Dh2 = "TCTCTGTGCCTATACGGGGGCACAATGATT"
        
        Dh1.kmers = sapply(1:nchar(Dh1),function(i)kmer.generator(Dh1,largo = i))
        Dh2.kmers = sapply(1:nchar(Dh2),function(i)kmer.generator(Dh2,largo = i))
        Dh1.matches = sapply(nchar(Dh1):1,function(i)sapply(1:length(Dh1.kmers[[i]]),function(j)regexpr(Dh1.kmers[[i]][j],df.final$CDR3.C.to.F)))
        Dh2.matches = sapply(nchar(Dh2):1,function(i)sapply(1:length(Dh2.kmers[[i]]),function(j)regexpr(Dh2.kmers[[i]][j],df.final$CDR3.C.to.F)))
        
        Dh1.indexes = lapply(Dh1.matches,function(x)apply(x,1,function(y)if(sum(y!= -1) ==1) y[y!= -1]else 0))
        Dh2.indexes = lapply(Dh2.matches,function(x)apply(x,1,function(y)if(sum(y!= -1) ==1) y[y!= -1]else 0))
        names(Dh1.indexes) = nchar(Dh1):1
        names(Dh2.indexes) = nchar(Dh2):1
        
        
        D1.start = rep(-1,length(df.final$CDR3.C.to.F))
        D1.length = D1.start
        D1.end = D1.start
        for (i in 1:(length(Dh1.indexes)-6))D1.start[which(Dh1.indexes[[i]]!=0)] = Dh1.indexes[[i]][which(Dh1.indexes[[i]]!=0)]
        for (i in 1:(length(Dh1.indexes)-6))D1.length[which(Dh1.indexes[[i]]!=0)] = as.numeric(names(Dh1.indexes[i]))
        D1.end = D1.length+D1.start-1
        D1.end[D1.end == -3] = -1
        
        D2.start = rep(-1,length(df.final$CDR3.C.to.F))
        D2.length = D2.start
        D2.end = D2.start
        for (i in 1:(length(Dh2.indexes)-6))D2.start[which(Dh2.indexes[[i]]!=0)] = Dh2.indexes[[i]][which(Dh2.indexes[[i]]!=0)]
        for (i in 1:(length(Dh2.indexes)-6))D2.length[which(Dh2.indexes[[i]]!=0)] = as.numeric(names(Dh2.indexes[i]))
        D2.end = D2.length+D2.start-1
        D2.end[D2.end == -3] = -1
        
        VD.i1 = D1.start-as.numeric(unlist(df.final$V.end))-1 
        VD.i1[D1.start== -1] = NA
        #barplot(table(VD.i1))
        DJ.i1 = as.numeric(df.final$J.start) - D1.end -1
        DJ.i1[D1.end== -1] = NA
        
        VD.i2 = D2.start-as.numeric(unlist(df.final$V.end))-1
        VD.i2[D2.start== -1] = NA
        #barplot(table(VD.i2))
        DJ.i2 = as.numeric(df.final$J.start) - D2.end -1
        DJ.i2[D2.end== -1] = NA
        DD.insertions = D2.start-D1.end-1
        DD.insertions[D2.start == -1 | D1.end == -1] = NA
        
        df.for.tcR = data.frame("Umi.count" =df.final$UMI.count , "Umi.proportion" =df.final$UMI.count/df.umi.sum , "Read.count" = df.final$Total.Freq , "Read.proportion" =  df.final$Total.Freq/df.Freq.sum , "CDR3.nucleotide.sequence" = df.final$CDR3.C.to.F , "CDR3.amino.acid.sequence" = df.final$CDR3.aa , "V.gene" = as.character(df.final$V.segment) , "J.gene" = as.character(df.final$J.segment), "D.gene" = rep("", length(df.final$UMI)) , "V.end" = df.final$V.end , "J.start" = df.final$J.start, "D5_1.end" =  D1.start, "D1_3.end" =  D1.end ,"D2_5.end" =  D2.start, "D2_3.end" =  D2.end , "VD.insertions" = VD.i1 , "DJ.insertions" = DJ.i1 , "Total.insertions" = as.numeric(unlist(df.final$J.start -df.final$V.end -1)) ,stringsAsFactors = FALSE)  
        return(df.for.tcR)
      }
      
      
    }
  }
  
  
  #####
  
  
  # Repertoire functions
  {
    UMI.extractor = function(sequenceR1,sequenceR2,umi.threshold = 1,mm=0,short_UPM="TGGTATCAACGCAGAGT",full_UPM_UMI="TGGTATCAACGCAGAGT[A-Z]{4}T[A-Z]{4}T[A-Z]{4}TCTTGGG"){
      #takes sread object as inputs, allows mm mismatch in the search for multiple patters for removal, but only exact match for UMI pattern recognition
      largo = length(sequenceR2)
      inicio = Sys.time()
      cat("UMI.extractor", sep = "")
      count_UPM = vcountPattern(short_UPM,sequenceR2,max.mismatch = mm)
      find_UPM = regexpr(full_UPM_UMI,sequenceR2[count_UPM == 1]) #i look twice because I need the count from gregexpr and regmatches is too slow with gregexpr
      umis = subseq(regmatches(sequenceR2[count_UPM == 1],find_UPM),18,31)
      filtro = (1:largo)[count_UPM==1][find_UPM!=-1]
      umi.filter = as.numeric(table(umis)[umis])>umi.threshold
      filtro = filtro[umi.filter]
      cat("  --  Total:",largo," --  UPM count:",sum(count_UPM==1)," --  UMI structure:",length(umi.filter), " --  UMI threshold: ",sum (umi.filter) , " --  Final filtering:", round(sum (umi.filter)/largo*100,1),"%\n")
      print (round(difftime(Sys.time(),inicio),2))
      return = list(R1 = sequenceR1[filtro],R2 = sequenceR2[filtro],filter = filtro,UMI = umis[umi.filter],stats = list(Total = largo,UPM_count=table(count_UPM),UMI_structure=length(umi.filter), UMI.threshold = sum (umi.filter),Time = round(difftime(Sys.time(),inicio),2)))
    }
    
    CDR3.extractor <- function(step1.1,C.rc.pattern = "GGAGCT[AG]TGAC",start.of.triplet = 10){
      #Using the R1 sequence from UMI.extractor finds the C region provided (as reverseComplement) and creates a reading frame matrix
      cat("\n",sep="")
      cat("CDR3.extractor",sep="")
      inicio=Sys.time()
      pattern_match = regexpr(C.rc.pattern,step1.1)
      rfs.1 = as.numeric(nchar(step1.1)-pattern_match)+start.of.triplet+1
      rfs.2 = rfs.1%%3 + 1
      resultados = c()
      
      filter.CDR3.extractor = pattern_match != -1
      ancho = (floor(nchar(step1.1[1])/3)-1)*3 # I'm assuming all sequences of equal length here
      
      RFj = suppressWarnings(translate(subseq(reverseComplement(step1.1[filter.CDR3.extractor]),rfs.2[filter.CDR3.extractor],width = ancho),if.fuzzy.codon = "solve" , no.init.codon = TRUE))
      cat ("  --  Sequences with C pattern: ",sum(filter.CDR3.extractor),"  (",round(sum(filter.CDR3.extractor)*100/length(pattern_match),1),"%)\n" , sep = "")
      print(round(difftime(Sys.time(),inicio,units = "auto"),digits = 2))
      return (list(translated.reads=RFj,filter.CDR3.extractor=filter.CDR3.extractor,reading.frames=rfs.2[filter.CDR3.extractor]))
    }
    
    Canonical.CDR3.j.trimmer <- function(filtered.translated.reads, filtered.r1.reads , reading.frames.table, CDR3.consensus, J.consensus, post.J = 27){
      # using a protein C and J consensus and the reading frame from CDR3.extractor trims the CDR3 region
      cat("\nCDR3 trimmer", sep ="")
      rf.orders = as.numeric(names(sort(table(reading.frames.table),decreasing = T)))
      J.match = regexpr(J.consensus,filtered.translated.reads)
      filtered.r1.reads.aa =tr.DNA3(filtered.r1.reads[J.match!=-1])
      CDR3.pos.1 = regexpr(CDR3.consensus,filtered.r1.reads.aa[[rf.orders[1]]])
      CDR3.pos.1 [CDR3.pos.1 >  J.match[J.match!=-1]] = -1
      CDR3.pos.2 = regexpr(CDR3.consensus,filtered.r1.reads.aa[[rf.orders[2]]][CDR3.pos.1==-1])
      CDR3.pos.2 [CDR3.pos.2 >  J.match[J.match!=-1][CDR3.pos.1==-1]] = -1
      CDR3.pos.3 = regexpr(CDR3.consensus,filtered.r1.reads.aa[[rf.orders[3]]][CDR3.pos.1==-1][CDR3.pos.2==-1])
      CDR3.pos.3 [CDR3.pos.3 >  J.match[J.match!=-1][CDR3.pos.1==-1][CDR3.pos.2==-1]] = -1
      CDR3.position.list = list(CDR3.pos.1,CDR3.pos.2,CDR3.pos.3)
      
      filter.list = list(which(J.match!=-1),which(J.match!=-1)[CDR3.pos.1==-1],which(J.match!=-1)[CDR3.pos.1==-1][CDR3.pos.2==-1])
      
      rf1= lapply(1:3,function(i)filter.list[[i]][ CDR3.position.list[[i]]  !=-1 & reading.frames.table[filter.list[[i]]] == ((rf.orders-1)%%3+1)[i]])
      rf2= lapply(1:3,function(i)filter.list[[i]][ CDR3.position.list[[i]]  !=-1 & reading.frames.table[filter.list[[i]]] == ((rf.orders+0)%%3+1)[i]])
      rf3= lapply(1:3,function(i)filter.list[[i]][ CDR3.position.list[[i]]  !=-1 & reading.frames.table[filter.list[[i]]] == ((rf.orders+1)%%3+1)[i]])
      {
        #each rf gives the position of CDR3 consensus in seq (as aminoacid) whithout the need to add filters
        #for example if we translate all reading frames for tr.DNA3(filtered.r1.reads[rf1[[1]]])
        #we will see the sequences where the consensus is in the reading frame 1 and also the J and C are in reading frame 1
        #in all the other lists only the CDR3 consensus is in the right frame (in rf2[[3]] for example, 
        #CDR3 is in the right reading frame in the third position of the list, while C and J are not,
        #but are 1 nucleotide away from each other in all 3 elements of the list because of this command(rf.orders%%3+1))
      }
      cdr1 = lapply(1:3,function(i)CDR3.position.list[[i]][ CDR3.position.list[[i]]  !=-1 & reading.frames.table[filter.list[[i]]] == ((rf.orders-1)%%3+1)[i]])
      cdr2 = lapply(1:3,function(i)CDR3.position.list[[i]][ CDR3.position.list[[i]]  !=-1 & reading.frames.table[filter.list[[i]]] == ((rf.orders+0)%%3+1)[i]])
      cdr3 = lapply(1:3,function(i)CDR3.position.list[[i]][ CDR3.position.list[[i]]  !=-1 & reading.frames.table[filter.list[[i]]] == ((rf.orders+1)%%3+1)[i]])
      
      CDR3s.1 = lapply(1:3,function(i)subseq(subseq(filtered.r1.reads[rf1[[i]]],rf.orders[i]),cdr1[[i]]*3+13,(J.match[rf1[[i]]]*3+post.J)))
      CDR3s.2 = lapply(1:3,function(i)subseq(subseq(filtered.r1.reads[rf2[[i]]],rf.orders[i]),cdr2[[i]]*3+13,(J.match[rf2[[i]]]*3+((rf.orders+0)%%3+1-rf.orders)[i]+post.J)))
      CDR3s.3 = lapply(1:3,function(i)subseq(subseq(filtered.r1.reads[rf3[[i]]],rf.orders[i]),cdr3[[i]]*3+13,(J.match[rf3[[i]]]*3+((rf.orders+1)%%3+1-rf.orders)[i]+post.J)))
      
      canonical.filter.list = list(sort(unlist(rf1)),sort(unlist(rf2)),sort(unlist(rf3)))
      CDR3s.list = list(do.call("c",CDR3s.1)[order(unlist(rf1))],do.call("c",CDR3s.2)[order(unlist(rf2))],do.call("c",CDR3s.3)[order(unlist(rf3))])
      
      largo = length(filtered.translated.reads)
      kept = length(unlist(canonical.filter.list))
      cat ("  --  Found canonical CDR3s ",kept," (",round(100*kept/largo,1), "%)  --  In frame ", round(100*length(canonical.filter.list[[1]])/kept,1), "%\n" ,sep = "")
      return(list(CDR3s=CDR3s.list,CDR3.filter=canonical.filter.list))
    }
    
    filter_V_MF.mk3 = function (step1.R2,tb.from=1, tb.to=3, tolerance=0,pat.start = 1, pat.end =110, database=MF_Vas.5){
      
      allele.names = names(database)
      MF.V.patterns = subseq(unlist(database),start = pat.start, end = pat.end)
      cat("Minimum pattern distance:",min(stringDist(MF.V.patterns,method = "hamming"))," --- ","Tolerance:",tolerance,sep = "")
      MF.dict.patterns = DNAStringSet(MF.V.patterns)
      MF.dict <- PDict(MF.dict.patterns,tb.start=tb.from,tb.end=tb.to)
      V.matrix <- vcountPDict (MF.dict, step1.R2, max.mismatch=tolerance)
      res.dicc = which(V.matrix>0,arr.ind = TRUE)
      V.matrix.split = split(res.dicc [,1],res.dicc [,2])
      resultado1 = lapply(V.matrix.split,function(x)paste0(unique(allele.names[as.numeric(x)]),collapse = ","))
      resultado = rep(0,length(step1.R2))
      resultado[as.numeric(names(resultado1))] = resultado1
      resultado.final = unlist(resultado)
      return(Vas = resultado.final)}
    
    V.call.x3 = function(query_start = 1 , query_end ,tol = 0, step1, CDR3e, filter_UMI_CDR3, R2_trim_start = 70, trim_length = 180, ambiguous = FALSE, database.V = MF_Vas.5, R2 = T){
      if (R2)trim_R2 = subseq(step1$R2[CDR3e$filter.CDR3.extractor][filter_UMI_CDR3][Vs.x == 0], R2_trim_start ,width = trim_length)else{
        trim_R2 = subseq(reverseComplement(step1$R1[CDR3e$filter.CDR3.extractor][filter_UMI_CDR3][Vs.x == 0]), R2_trim_start ,width = trim_length)
      }
      
      target = DNAStringSet(names(table(trim_R2)))
      temp.res = filter_V_MF.mk3(target ,pat.start = query_start, pat.end = query_end, tolerance = tol, database = database.V)
      temp.V.call = temp.res[match(trim_R2,target)]
      cat(" -- found: ",mean(temp.V.call != 0))
      if (ambiguous == FALSE){
        temp.V.call[grepl(",",temp.V.call)] = 0
        cat(" -- kept: ",mean(temp.V.call != 0))}
      Vs.x[Vs.x == 0]<<-temp.V.call
      cat(" -- overall:",mean(Vs.x != 0),"\n")
      
    }
    
    fast.table.parser.v1 = function(step1.x,CDR3e.x,Ctr.x,filter_UMI_CDR3_x,Vxs,threshold = 2){
      # threshold is the minimum count at which UMI-CDR3s-Vs combinations are kept (should be selected based on read depth and sample size)
      #
      # threshold.single is the threshold below which single UMIs or CDR3s are eliminated to make the function run faster
      
      # Make an index for sample length
      
      # Make data frames for each sample
      frames.x.UMIs = as.character(step1.x$UMI[CDR3e.x$filter.CDR3.extractor][filter_UMI_CDR3_x])
      frames.x.CDR3s = as.character(do.call("c",Ctr.x$CDR3s)[do.call("c",Ctr.x$CDR3.filter)  %in% filter_UMI_CDR3_x])
      frames.x.Vs = as.character(Vxs)
      # Apply and combine threshold.single filter
      
      # Combine and remove columns where the count is less than 2 for either
      frames.umi.cdr3.x.df = split(data.frame(UMIs = frames.x.UMIs, CDR3s = frames.x.CDR3s), frames.x.Vs)
      
      # Count UMI-CDR3-Vs combinations and keep only above threshold (which should be selected based on read depth)
      frames.umi.count.x = lapply(frames.umi.cdr3.x.df[-1],function(x)separate(data.frame(table(paste(x[,1],x[,2],sep = ","))),1,into = c("UMI","CDR3")))
      frames.umi.count.x.2 = frames.umi.count.x[sapply(frames.umi.count.x,nrow)>0]
      frames.umi.count.x.3 = lapply(frames.umi.count.x.2,function(x)x[x[,3]>=threshold,])
      frames.umi.count.x.4 = frames.umi.count.x.3[sapply(frames.umi.count.x.3,nrow)>0]
      frames.umi.count.x.5 = lapply(1:length(frames.umi.count.x.4),function(i)cbind(frames.umi.count.x.4[[i]],V.segment=names(frames.umi.count.x.4)[i]))
      cat("Duplicated UMI+CDR3s per sample: " , sum(table(unlist(lapply(frames.umi.count.x.5,function(x)paste(paste(x[,1],x[,2],sep = ","),x[,4],sep = ","))))!=1))
      
      return (frames.umi.count.x.5)
    }
    
    error.correction.mk5 =  function(secuencias) {
      retain = c()
      discard= list()
      if (nrow(secuencias)==1)retain = rbind(retain,secuencias) else{
        nf = nodes.finder.UMI(secuencias[,1])
        if(sum(nf)==0) retain = rbind(retain,secuencias) else{
          target.umi = DNAStringSet(secuencias[,1])[nf] # this assumes that there won't be many errors unconnected, but it might not be a safe assumption for CDR3s (it probably is for UMIs)
          vD.umi=vcountPDict(PDict(target.umi,tb.start = 5,tb.end = 5),target.umi,max.mismatch = 1)
          target.cdr3 = DNAStringSet(secuencias[,2])[nf]
          vD.cdr3 =vcountPDict(PDict(target.cdr3 ,tb.start = 1,tb.end = 1),target.cdr3 ,max.mismatch = 2)
          vD.joint = vD.umi+vD.cdr3
          res = which(vD.joint==2,arr.ind = TRUE)
          edges = res[res[,1]!=res[,2],]
          if (length(edges)==0) retain = rbind(retain,secuencias) else{
            gr = graph(as.vector(t(edges)))
            cls=clusters(gr)
            cls.list = lapply(seq_along(cls$csize)[cls$csize > 1], function(x) V(gr)[cls$membership %in% x])
            pre.sel = lapply(cls.list,function(clu)secuencias[(nf),][clu,])
            sel = do.call(rbind,lapply(pre.sel,function(x)x[which.max(x$Freq),]))
            discard = append(discard,pre.sel)
            retain = rbind(retain,sel)
            retain = rbind(retain,secuencias[!(nf),])
            retain = rbind(retain,secuencias[nf,][!(1:nrow(secuencias[nf,])%in%(unlist(cls.list))),])
          }
        }
      }
      return (list(corrected.list = retain,errors = discard))
    }
    
    nodes.finder.UMI = function(secuencia){
      secuencia=as.character(secuencia)
      oR = lapply(1:14,function(i)paste0(subseq(secuencia,1,width = (i-1)),subseq(secuencia,i+1,width = (14-i))))
      nodos= rowSums(sapply(oR,function(x)duplicated(x)|duplicated(x,fromLast = TRUE)))>0
      return(nodos)
    }
    graph.builder = function(nodes){
      conflicting.UMIs.list = list()
      if (length(nodes!=0)){
        c = as.vector(t(nodes))
        g <- graph(c)
        cl <- clusters(g)
        conflicting.UMIs.list <- lapply(seq_along(cl$csize)[cl$csize > 1], function(x) V(g)[cl$membership %in% x])
      } else{conflicting.UMIs.list=NULL}
      return (conflicting.UMIs.list)
    }
    
    matrix.builder = function(distancia){
      matriz = as.matrix(distancia)
      matriz[upper.tri(matriz)] <- NA
      diag(matriz) = NA
      return(matriz)  
    }
    kmer.generator = function(secuencias,largo)lapply(as.character(secuencias),function(x)sapply(1:(nchar(x)-(largo-1)),function(y)subseq(x,y,width = largo)))
    
    df.reparser.MF = function(df,remove.duplicates = T,aggregate.umis = F){
      initial = nrow(df)
      df = df[order(df$Umi.count, decreasing = T),]                 #order in decreasing UMI count
      df = df[df$J.start-1<nchar(df$CDR3.nucleotide.sequence),]     #remove Js assigned out of CDR3
      df = df[df$J.gene!="" & df$J.gene!=0 ,]                       #remove nonassigned Js                                  
      df = df[df$V.end >0 & df$J.start>0,]                          #remove negative V end or J start assignment
      df = df[!(grepl("[^A,C,G,T]",df$CDR3.nucleotide.sequence)),]  #remove N nucleotides
      df$V.end2 = df$V.end                                          #create a new end of V variable that is always less than start of J
      df$J.length = nchar(df$CDR3.nucleotide.sequence) - df$J.start +1 
      df$V.end2[df$V.end >= df$J.start] = df$J.start[df$V.end >= df$J.start]-1
      df$pc.nt = pc.generator.nt(df)
      if (aggregate.umis){df.umi.agg = tapply(df$Umi.count,df$pc.nt,sum)
      df$Umi.count = df.umi.agg[df$pc.nt]}
      if(remove.duplicates)df = df[!duplicated(df$pc.nt),]
      df$L = nchar(df$CDR3.nucleotide.sequence)
      df = df[shuffler.indexes.df(df),]
      df = df[duplicated(df$L)|duplicated(df$L,fromLast = T),]
      return(df)}
    
    pc.generator.nt = function(df)paste(df$CDR3.nucleotide.sequence,df$V.gene,df$J.gene,sep = ";")
    
    shuffler.indexes.df = function(df)as.numeric(rev(unlist(tapply((1:nrow(df))[order(as.numeric(df$Umi.count),decreasing = T)],sort(as.numeric(df$Umi.count),decreasing = T),function(x)sample(as.character(x))))))
    
    pc.calculator = function(df.list, unicas = F){
      if (is.null(names(df.list)))names(df.list) = 1:length(df.list)
      if(unicas) pc = lapply(df.list,pc.generator.nt.u) else pc = lapply(df.list,pc.generator.nt)
      publicity = as.numeric(table(unlist(pc))[unlist(pc)])
      pc.table = table(table(unlist(pc)))
      publicity.sp = split(publicity,unlist(sapply(1:length(pc),function(i)rep(names(pc)[i],length(pc[[i]])))))
      publicity.sp = publicity.sp[match(names(pc),names(publicity.sp))]
      return(list(pc=pc,publicity=publicity.sp,pc.table = pc.table))}
    
    df.2.publicity.calculator = function(df.list){
      df.list.pc = pc.calculator(df.list)
      for (i in 1:length(df.list))df.list[[i]]$pub = df.list.pc$publicity[[i]]
      return(df.list)}
  }
  
  #####
  {
    MF.tables.a = list()
    MF.tables.b = list()
    MF.tables.g = list()
    MF.tables.d = list()
    MF.tables.h = list()
    
    
    MF.errors.a = list()
    MF.errors.b = list()
    MF.errors.g = list()
    MF.errors.d = list()
    MF.errors.h = list()
    
    MF.stats.a = list()
    MF.stats.b = list()
    MF.stats.g = list()
    MF.stats.d = list()
    MF.stats.h = list()
    #####
    
    MF.files =list.files(path= path.MF.files ,pattern=".fastq.gz",recursive = TRUE)
    MF.files.2 =list.files(path= path.MF.files.2 ,pattern=".fastq.gz",recursive = TRUE)
    umi_cdr3_threshold = 3
    inicio = Sys.time()
  }
  {
    
    for(s in 1:10){
      
      MF.files.sel = MF.files[(s*2-1):(s*2)]
      MF.Fastq = mapply(function(x,y)assign(x,readFastq(y)),sub(".fastq.gz","",sub(".*/","",MF.files.sel)),paste0(path.MF.files,MF.files.sel)) #mapply from file name to path 
      MF.umi = UMI.extractor(sread(MF.Fastq[[1]]),sread(MF.Fastq[[2]]),umi.threshold = 2)
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern = "TGTAAATAGCTGGTTTAACCTC",start.of.triplet = 23)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.a, J.consensus.MF.a)
      
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 60, tol = 1, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 60, tol = 1, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        MF.Vs = sub("\\..*","",Vs.x)
      }
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.a[[s]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.a[[s]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.a[[s]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.a[[s]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.a[[s]]))
      
    }
    
    for(s in 11:20){
      
      MF.files.sel = MF.files[(s*2-1):(s*2)]
      MF.Fastq = mapply(function(x,y)assign(x,readFastq(y)),sub(".fastq.gz","",sub(".*/","",MF.files.sel)),paste0(path.MF.files,MF.files.sel)) #mapply from file name to path 
      MF.umi = UMI.extractor(sread(MF.Fastq[[1]]),sread(MF.Fastq[[2]]))
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern = "TGACATCAATACCTGG",start.of.triplet = 14)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.b, J.consensus.MF.b)
      
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 120, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4, trim_length = 230)
        V.call.x3(query_end = 80, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4, trim_length = 230)
        V.call.x3(query_end = 40, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4, trim_length = 230)
        MF.Vs = Vs.x
      }
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4,trim_length = 230,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4,trim_length = 230,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.b[[(s-10)]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.b[[(s-10)]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.b[[(s-10)]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.b[[s-10]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.b[[s-10]]))
      
    }
    
    for(s in 21:30){
      
      MF.files.sel = MF.files[(s*2-1):(s*2)]
      MF.Fastq = mapply(function(x,y)assign(x,readFastq(y)),sub(".fastq.gz","",sub(".*/","",MF.files.sel)),paste0(path.MF.files,MF.files.sel)) #mapply from file name to path 
      MF.umi = UMI.extractor(sread(MF.Fastq[[1]]),sread(MF.Fastq[[2]]))
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern = "ACACTTCAGGTTT",start.of.triplet = 11)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.g, J.consensus.MF.g)
      
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 120, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 230)
        V.call.x3(query_end = 80, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 230)
        V.call.x3(query_end = 40, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 230)
        MF.Vs = Vs.x
      }
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 250,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      print(mean(R2.R1.agreement))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.g[[(s-20)]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.g[[(s-20)]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.g[[(s-20)]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.g[[s-20]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.g[[s-20]]))
      
    }
    
    for(s in 31:40){
      
      MF.files.sel = MF.files[(s*2-1):(s*2)]
      MF.Fastq = mapply(function(x,y)assign(x,readFastq(y)),sub(".fastq.gz","",sub(".*/","",MF.files.sel)),paste0(path.MF.files,MF.files.sel)) #mapply from file name to path 
      MF.umi = UMI.extractor(sread(MF.Fastq[[1]]),sread(MF.Fastq[[2]]))
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern =  "TTTCAGGAACATC" ,start.of.triplet = 11)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.d, J.consensus.MF.d)
      
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 60, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 60, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        
      }
      
      MF.Vs = sub("\\..*","",Vs.x)
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 270, query_end = 320, tol = 5, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 250,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      print(mean(R2.R1.agreement))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.d[[(s-30)]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.d[[(s-30)]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.d[[(s-30)]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.d[[s-30]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.d[[s-30]]))
      
    }
    
    for(s in 41:50){
      
      MF.files.sel = MF.files[(s*2-1):(s*2)]
      MF.Fastq = mapply(function(x,y)assign(x,readFastq(y)),sub(".fastq.gz","",sub(".*/","",MF.files.sel)),paste0(path.MF.files,MF.files.sel)) #mapply from file name to path 
      MF.umi = UMI.extractor(sread(MF.Fastq[[1]]),sread(MF.Fastq[[2]]))
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern =  "GTCCAGATGGTGCG" ,start.of.triplet = 14)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.h, J.consensus.MF.h)
      
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 130, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 230)
        V.call.x3(query_end = 80, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 230)
        V.call.x3(query_end = 60, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 230)
      }
      
      MF.Vs = Vs.x
      
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 270, query_end = 320, tol = 5, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 250,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      print(mean(R2.R1.agreement))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.h[[(s-40)]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.h[[(s-40)]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.h[[(s-40)]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.h[[s-40]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.h[[s-40]]))
      
    }
    
    for(s in 1:5){
      
      MF.files.sel = MF.files.2[grepl("MF",MF.files.2)][(s*2-1):(s*2)]
      MF.Fastq = mapply(function(x,y)assign(x,readFastq(y)),sub(".fastq.gz","",sub(".*/","",MF.files.sel)),paste0(path.MF.files.2,MF.files.sel)) #mapply from file name to path 
      MF.umi = UMI.extractor(sread(MF.Fastq[[1]]),sread(MF.Fastq[[2]]))
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern = "TGTAAATAGCTGGTTTAACCTC",start.of.triplet = 23)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.a, J.consensus.MF.a)
      
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        MF.Vs = sub("\\..*","",Vs.x)
        
      }
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.a[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.a[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.a[[s+10]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.a[[s+10]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.a[[s+10]]))
      
      
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern = "TGACATCAATACCTGG",start.of.triplet = 14)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.b, J.consensus.MF.b)
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 120, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4,trim_length = 230)
        V.call.x3(query_end = 80, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4,trim_length = 230)
        V.call.x3(query_end = 40, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4,trim_length = 230)
        MF.Vs = Vs.x
      }
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4,trim_length = 230,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vbs.4,trim_length = 230,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.b[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.b[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.b[[s+10]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.b[[s+10]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.b[[s+10]]))
      
      # Gamma
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern = "ACACTTCAGGTTT",start.of.triplet = 11)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.g, J.consensus.MF.g)
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 120, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 230)
        V.call.x3(query_end = 80, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 230)
        V.call.x3(query_end = 40, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 230)
        MF.Vs = Vs.x
      }
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vg.2,trim_length = 250,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      print(mean(R2.R1.agreement))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      if(sum(MF.Vs!=0)>20){
        MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 2)
        MF.ec = lapply(MF.ftp,error.correction.mk5)
        MF.tables.g[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
        MF.errors.g[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      }
      # Delta
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern =  "TTTCAGGAACATC" ,start.of.triplet = 11)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.d, J.consensus.MF.d)
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 230)
        V.call.x3(query_end = 110, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.Vas.observed,trim_length = 230)
        MF.Vs = sub("\\..*","",Vs.x)
      }
      
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 270, query_end = 320, tol = 5, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF_Vas.5,trim_length = 250,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      print(mean(R2.R1.agreement))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      if(sum(MF.Vs!=0)>20){
        MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
        MF.ec = lapply(MF.ftp,error.correction.mk5)
        MF.tables.d[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
        MF.errors.d[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
        MF.stats.d[[s+10]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.g[[s+10]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.d[[s+10]]))
        
      }
      # IgH
      
      CDR3.e = CDR3.extractor(MF.umi$R1,C.rc.pattern =  "GTCCAGATGGTGCG" ,start.of.triplet = 14)
      CDR3.tr = Canonical.CDR3.j.trimmer(CDR3.e$translated.reads,reverseComplement(MF.umi$R1[CDR3.e$filter.CDR3.extractor]),CDR3.e$reading.frames, CDR3.consensus.MF.h, J.consensus.MF.h)
      CDR3.tr_cdr3s = do.call(c,CDR3.tr$CDR3s)
      CDR3.tr_fil = do.call(c,CDR3.tr$CDR3.filter)
      UMI_CDR3 = paste(MF.umi$UMI[CDR3.e$filter.CDR3.extractor][CDR3.tr_fil],CDR3.tr_cdr3s) 
      filter_UMI_CDR3 = CDR3.tr_fil[as.logical(table(UMI_CDR3)[UMI_CDR3]  >= umi_cdr3_threshold)]
      
      {
        Vs.x = rep(0,length(filter_UMI_CDR3))
        V.call.x3(query_end = 130, tol = 0, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 230)
        V.call.x3(query_end = 80, tol = 2, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 230)
        V.call.x3(query_end = 60, tol = 4, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 230)
        
      }
      MF.Vs = Vs.x
      
      ###
      Vs.x = rep(0,length(filter_UMI_CDR3))
      V.call.x3(R2_trim_start = 10,query_start = 240, query_end = 300, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 260, query_end = 320, tol = 3, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 250,ambiguous = T,R2 = F)
      V.call.x3(R2_trim_start = 10,query_start = 270, query_end = 320, tol = 5, step1 = MF.umi,CDR3e =  CDR3.e,filter_UMI_CDR3 = filter_UMI_CDR3 ,database.V = MF.IgHs,trim_length = 250,ambiguous = T,R2 = F)
      MF.Vs.r1 = Vs.x
      R2.R1.agreement= mapply(function(x,y)grepl(x,y),MF.Vs,MF.Vs.r1)
      print(mean(R2.R1.agreement[MF.Vs!=0&MF.Vs.r1!=0]))
      print(mean(R2.R1.agreement))
      MF.Vs[!R2.R1.agreement] = 0
      ###
      
      MF.ftp = fast.table.parser.v1(MF.umi,CDR3.e,CDR3.tr,filter_UMI_CDR3,MF.Vs,threshold = 3)
      MF.ec = lapply(MF.ftp,error.correction.mk5)
      MF.tables.h[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)x[[1]]))
      MF.errors.h[[s+10]]=do.call(rbind,lapply(MF.ec,function(x)do.call(rbind,x[[2]])))
      MF.stats.h[[s+10]] = c(MF.umi$stats,CDR3.e = sum(CDR3.e$filter.CDR3.extractor), Canonical_CDR3 = length(CDR3.tr_fil), Canonical_UMI_thr = length(filter_UMI_CDR3) ,Corrected = sum(MF.tables.h[[s+10]][,3]),UMI_CDR3s = length(unique(UMI_CDR3[filter_UMI_CDR3])), Clones = nrow(MF.tables.h[[s+10]]))
      
    }
  }
  sapply(1:10,function(i)sum(table(paste(MF.tables.a[[i]][,1],MF.tables.a[[i]][,2]))>1))
  sapply(1:10,function(i)sum(table(paste(MF.tables.d[[i]][,1],MF.tables.d[[i]][,2]))>1))
  sapply(1:10,function(i)sum(table(paste(MF.tables.g[[i]][,1],MF.tables.g[[i]][,2]))>1))
  sapply(1:10,function(i)sum(table(paste(MF.tables.d[[i]][,1],MF.tables.d[[i]][,2]))>1))
  sapply(1:10,function(i)sum(table(paste(MF.tables.h[[i]][,1],MF.tables.d[[i]][,2]))>1))
  
  MF.a.parsed = lapply(1:5,function(i)MF.TA.data.parser(do.call(rbind,MF.tables.a[c(i,i+5,i+10)])))
  MF.b.parsed = lapply(1:5,function(i)MF.TB.data.parser.v2(do.call(rbind,MF.tables.b[c(i,i+5,i+10)])))
  MF.g.parsed = lapply(1:5,function(i)MF.TG.data.parser(do.call(rbind,MF.tables.g[c(i,i+5,i+10)])))
  MF.d.parsed = lapply(1:5,function(i)MF.TD.data.parser(do.call(rbind,MF.tables.d[c(i,i+5,i+10)])))
  MF.h.parsed = lapply(1:5,function(i)MF.IgH.data.parser(do.call(rbind,MF.tables.h[c(i,i+5,i+10)])))
  
  MF.x.list = list(MF.a.parsed,MF.b.parsed,MF.g.parsed,MF.d.parsed,MF.h.parsed)
  MF.x.list.2 = lapply(MF.x.list,function(x)lapply(x[-1],df.reparser.MF)) #remove fish 1
  MF.x.list.3 = lapply(MF.x.list.2,df.2.publicity.calculator)
  MF.list = MF.x.list.3
  print(difftime(Sys.time(),inicio))
  save(MF.list,file = "MF.list")
  save(MF_Vas.5,file = "MF_Vas.5")
  
}

